<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<HTML>
<HEAD>
<META HTTP-EQUIV="Content-Type" CONTENT="text/html; charset=iso-8859-1">
<META name="keywords" content="BLAST Parameters">
<TITLE>WU&nbsp;BLAST 2.0 Parameter Descriptions</TITLE>
<LINK media="screen" href="styles/blast_styles.css" type="text/css" rel="stylesheet">
</HEAD>
<BODY>

<H2>WU&nbsp;<A HREF="http://blast.wustl.edu">BLAST</A>&nbsp;2.0 Parameters</H2>

<H3>Command Line Options</H3>
<!-- Remaining:  ckstdlk, summethod, scalematrix, vdbdescmax -->

<P>
The complete list of command line options supported by
<A HREF="http://blast.wustl.edu">WU-BLAST&nbsp;2.0</A> is provided in the
<A HREF="#table1">tables</A> below.
The information provided here
comprises their definitive description.
This information should be considered valid <I>only</I>
for the current (most recent) version of the software.
If you find an inconsistency between the advertised
behavior and the actual behavior of the software,
first be sure you are using the latest version,
as indicated by the date of the latest release shown at
<A HREF="http://blast.wustl.edu" TARGET="_blank">
http://blast.wustl.edu</A>.
If the inconsistency persists after upgrading,
please report it to
<script language="JavaScript" type="text/javascript">
<!--
	var irst = 'ma';
	var second = 'il';
	var third = 'to:';
	var ddress = '&#87;&#97;&#114;&#114;&#101;&#110;&#32;&#71;&#105;&#115;&#104;&#32;&#60;&#103;&#105;&#115;&#104;';
	var domain = '&#119;&#97;&#116;&#115;&#111;&#110;&#46;&#119;&#117;&#115;&#116;&#108;&#46;&#101;&#100;&#117;&#62;&#63;&#115;&#117;&#98;&#106;&#101;&#99;&#116;&#61;&#66;&#76;&#65;&#83;&#84;&#32;&#69;&#45;&#109;&#97;&#105;&#108;';
	document.write('<a href="');
	document.write(irst+second+third);
	document.write(ddress);
	document.write('&#64;');
	document.write(domain);
	document.write('" title="Click here to compose an e-mail message">');
	document.write('BLAST E-mail<\/a>.');
// -->
</script>
If you wish to continue using an older version of the software
instead of upgrading,
please consult the copy of <CODE>parameters.html</CODE> that came bundled with that version;
it may be more accurate for your purposes than the on-line documentation.
Where differences arise between the bundled file and the on-line version,
they may be due to improvements or corrections that have been made
to the documentation itself
or due to actual differences between versions of the software.
</P>

<P>
When this web page can not be conveniently accessed,
terse descriptions of most items may be obtained by entering the relevant
BLAST program name on the command line without any arguments.
The most recent version of the page you are viewing is located
<A HREF="http://blast.wustl.edu/blast/parameters.html">here</A>.
A PDF version of the page is available
<A HREF="http://blast.wustl.edu/blast/parameters.pdf">here</A>.
For most of the options, a logical
<A HREF="http://blast.wustl.edu/doc/WU-BLAST-ControlParms.pdf">
diagram</A>
indicates where each has its effect.
</P>

<P>
Command line options for the obsolete WU- and NCBI-BLAST version 1.4,
first released in 1994,
often apply unchanged to WU-BLAST 2.0,
which yields a high degree of upward compatibility.
While <A HREF="http://blast.wustl.edu/blast-1.4">BLAST&nbsp;1.4</A> is many years old now,
if you are interested in it,
<I>e.g.,</I> for reasons of reproducing prior results,
please see the
<A HREF="http://blast.wustl.edu/blast-1.4/blast1.pdf">BLAST&nbsp;1.4 manual page</A>
in PDF format.
</P>


<H3>Command Line Syntax</H3>
<P>
Aside from the first two command line arguments
(database name and query filename), which are required items,
the WU-BLAST search programs support a flexible syntax
for command line options and parameters.
Parsing of the command line is generally alphabetical case-<I>in</I>dependent.
A leading hyphen (-) is unnecessary on option names
but may improve its human readability.
Parameter values can optionally be specified using an equals sign (=).
Combined use of hyphens and equals signs is allowed and
does not need to be consistently applied
throughout a given command line.
Large integer values can be specified using floating point representation
(<I>e.g.</I>, 1e9 instead of 1000000000).
For parameters with single letter names, neither a hyphen nor an equals
sign is necessary.
</P>

<P>
The basic command line syntax is:
</P>

<PRE>
	<I>&lt;program&gt; &lt;database&gt; &lt;query&gt; [options...]</I>
</PRE>

<P>
where <I>&lt;program&gt;</I> is one of
<CODE>blastp</CODE>,
<CODE>blastn</CODE>,
<CODE>blastx</CODE>,
<CODE>tblastn</CODE>
and
<CODE>tblastx</CODE>;
<I>&lt;database&gt;</I>
is the name of the database to search
(previously formatted with <CODE>xdformat</CODE>);
<I>&lt;query&gt;</I> is the name of a file containing one or more query
sequences in FASTA format;
and <I>[options...]</I> is a list of zero or more command line options
and parameter settings.
</P>


<P>
As examples of the command line flexibility available,
each of the following command lines are valid and equivalent:
</P>
<PRE>
      blastp nr myquery.aa  v=10  b=100  filter=seg    e=1e-10    nogaps
      blastp nr myquery.aa  V=10  B=100  filter=seg    E=1e-10    nogaps
      blastp nr myquery.aa -V=10 -B=100 -filter=seg   -E=1e-10   -nogaps
      blastp nr myquery.aa -V10  -B100  -filter seg   -E1e-10    -nogaps
      blastp nr myquery.aa -V10  -B100  -filter "seg" -E1e-10    -nogaps
      blastp nr myquery.aa -V 10 -B 100 -filter seg   -E 1e-10   -nogaps
      blastp nr myquery.aa  V 10  B 100  filter seg    E 1e-10    nogaps
      blastp nr myquery.aa  -v10  B=100  FILTER=seg   -e=1e-10   -nogaps
</PRE>

<P STYLE="page-break-before: always">
</P>

<A NAME="table1"></A>
<!-- TABLE_INSERT -->
<!-- TABLE_START -->
<H3>Table of Options</H3>
<TABLE BORDER="1" CELLPADDING="4">
<TR>
<TD><A HREF="#altscore">altscore</A></TD>
<TD><A HREF="#e">E</A></TD>
<TD><A HREF="#gapw">gapW</A></TD>
<TD><A HREF="#lcfilter">lcfilter</A></TD>
<TD><A HREF="#nwstart">nwstart</A></TD>
<TD><A HREF="#qtype">qtype</A></TD>
<TD><A HREF="#stats">stats</A></TD>
</TR>
<TR>
<TD><A HREF="#b">B</A></TD>
<TD><A HREF="#e2">E2</A></TD>
<TD><A HREF="#gapx">gapX</A></TD>
<TD><A HREF="#lcmask">lcmask</A></TD>
<TD><A HREF="#o">O</A></TD>
<TD><A HREF="#r">R</A></TD>
<TD><A HREF="#sump">sump</A></TD>
</TR>
<TR>
<TD><A HREF="#bottom">bottom</A></TD>
<TD><A HREF="#echofilter">echofilter</A></TD>
<TD><A HREF="#getenv">getenv</A></TD>
<TD><A HREF="#links">links</A></TD>
<TD><A HREF="#olfraction">olfraction</A></TD>
<TD><A HREF="#restest">restest</A></TD>
<TD><A HREF="#t">T</A></TD>
</TR>
<TR>
<TD><A HREF="#c">C</A></TD>
<TD><A HREF="#endgetenv">endgetenv</A></TD>
<TD><A HREF="#gi">gi</A></TD>
<TD><A HREF="#m">M</A></TD>
<TD><A HREF="#olmax">olmax</A></TD>
<TD><A HREF="#s">S</A></TD>
<TD><A HREF="#top">top</A></TD>
</TR>
<TR>
<TD><A HREF="#cdb">cdb</A></TD>
<TD><A HREF="#endputenv">endputenv</A></TD>
<TD><A HREF="#globalexit">globalexit</A></TD>
<TD><A HREF="#maskextra">maskextra</A></TD>
<TD><A HREF="#pingpong">pingpong</A></TD>
<TD><A HREF="#s2">S2</A></TD>
<TD><A HREF="#topcomboe">topcomboE</A></TD>
</TR>
<TR>
<TD><A HREF="#compat1.3">compat1.3</A></TD>
<TD><A HREF="#errors">errors</A></TD>
<TD><A HREF="#golfraction">golfraction</A></TD>
<TD><A HREF="#matrix">matrix</A></TD>
<TD><A HREF="#poissonp">poissonp</A></TD>
<TD><A HREF="#seqtest">seqtest</A></TD>
<TD><A HREF="#topcombon">topcomboN</A></TD>
</TR>
<TR>
<TD><A HREF="#compat1.4">compat1.4</A></TD>
<TD><A HREF="#evalues">evalues</A></TD>
<TD><A HREF="#golmax">golmax</A></TD>
<TD><A HREF="#mformat">mformat</A></TD>
<TD><A HREF="#postsw">postsw</A></TD>
<TD><A HREF="#shortqueryok">shortqueryok</A></TD>
<TD><A HREF="#ucdb">ucdb</A></TD>
</TR>
<TR>
<TD><A HREF="#consistency">consistency</A></TD>
<TD><A HREF="#filter">filter</A></TD>
<TD><A HREF="#gspmax">gspmax</A></TD>
<TD><A HREF="#mmio">mmio</A></TD>
<TD><A HREF="#progress">progress</A></TD>
<TD><A HREF="#soffset">soffset</A></TD>
<TD><A HREF="#v">V</A></TD>
</TR>
<TR>
<TD><A HREF="#cpus">cpus</A></TD>
<TD><A HREF="#gapall">gapall</A></TD>
<TD><A HREF="#h">H</A></TD>
<TD><A HREF="#msgstyle">msgstyle</A></TD>
<TD><A HREF="#prune">prune</A></TD>
<TD><A HREF="#sort_by_count">sort_by_count</A></TD>
<TD><A HREF="#w">W</A></TD>
</TR>
<TR>
<TD><A HREF="#ctxfactor">ctxfactor</A></TD>
<TD><A HREF="#gapdecayrate">gapdecayrate</A></TD>
<TD><A HREF="#haltonfatal">haltonfatal</A></TD>
<TD><A HREF="#n">N</A></TD>
<TD><A HREF="#putenv">putenv</A></TD>
<TD><A HREF="#sort_by_highscore">sort_by_highscore</A></TD>
<TD><A HREF="#warnings">warnings</A></TD>
</TR>
<TR>
<TD><A HREF="#dbbottom">dbbottom</A></TD>
<TD><A HREF="#gape">gapE</A></TD>
<TD><A HREF="#hitdist">hitdist</A></TD>
<TD><A HREF="#nogaps">nogaps</A></TD>
<TD><A HREF="#pvalues">pvalues</A></TD>
<TD><A HREF="#sort_by_pvalue">sort_by_pvalue</A></TD>
<TD><A HREF="#wink">wink</A></TD>
</TR>
<TR>
<TD><A HREF="#dbchunks">dbchunks</A></TD>
<TD><A HREF="#gape2">gapE2</A></TD>
<TD><A HREF="#hspmax">hspmax</A></TD>
<TD><A HREF="#nonnegok">nonnegok</A></TD>
<TD><A HREF="#q">Q</A></TD>
<TD><A HREF="#sort_by_subjectlength">sort_by_subjectlength</A></TD>
<TD><A HREF="#wordmask">wordmask</A></TD>
</TR>
<TR>
<TD><A HREF="#dbgcode">dbgcode</A></TD>
<TD><A HREF="#gaph">gapH</A></TD>
<TD><A HREF="#hspsepqmax">hspsepQmax</A></TD>
<TD><A HREF="#nosegs">nosegs</A></TD>
<TD><A HREF="#qframe">qframe</A></TD>
<TD><A HREF="#sort_by_totalscore">sort_by_totalscore</A></TD>
<TD><A HREF="#wstrict">wstrict</A></TD>
</TR>
<TR>
<TD><A HREF="#dbrecmax">dbrecmax</A></TD>
<TD><A HREF="#gapk">gapK</A></TD>
<TD><A HREF="#hspsepsmax">hspsepSmax</A></TD>
<TD><A HREF="#noseqs">noseqs</A></TD>
<TD><A HREF="#qoffset">qoffset</A></TD>
<TD><A HREF="#span">span</A></TD>
<TD><A HREF="#x">X</A></TD>
</TR>
<TR>
<TD><A HREF="#dbrecmin">dbrecmin</A></TD>
<TD><A HREF="#gapl">gapL</A></TD>
<TD><A HREF="#k">K</A></TD>
<TD><A HREF="#notes">notes</A></TD>
<TD><A HREF="#qrecmax">qrecmax</A></TD>
<TD><A HREF="#span1">span1</A></TD>
<TD><A HREF="#xmlcompact">xmlcompact</A></TD>
</TR>
<TR>
<TD><A HREF="#dbslice">dbslice</A></TD>
<TD><A HREF="#gaps">gaps</A></TD>
<TD><A HREF="#kap">kap</A></TD>
<TD><A HREF="#novalidctxok">novalidctxok</A></TD>
<TD><A HREF="#qrecmin">qrecmin</A></TD>
<TD><A HREF="#span2">span2</A></TD>
<TD><A HREF="#y">Y</A></TD>
</TR>
<TR>
<TD><A HREF="#dbtop">dbtop</A></TD>
<TD><A HREF="#gaps2">gapS2</A></TD>
<TD><A HREF="#l">L</A></TD>
<TD><A HREF="#nwlen">nwlen</A></TD>
<TD><A HREF="#qres">qres</A></TD>
<TD><A HREF="#spoutmax">spoutmax</A></TD>
<TD><A HREF="#z">Z</A></TD>
</TR>
</TABLE>
<!-- TABLE_END -->


<P STYLE="page-break-before: always">
</P>

<A NAME="table2"></A>
<H3>Table of Options with Descriptions</H3>

<TABLE BORDER="2" CELLPADDING="4">
	<TR>
		<TD>Option</TD>
		<TD>Description</TD>
	</TR>
	<TR>
		<TD>
		&nbsp;&nbsp;<CODE>altscore=</CODE>&quot;<I>score_spec</I>&quot;</TD>
		<TD><A NAME="altscore"></A>
<!-- PARM="altscore" -->
		alter individual scores or entire rows or columns
		of scores in a scoring matrix,
		without editing the scoring matrix file itself.
		<I>Score_spec</I> is a quoted character string
		consisting of three components, each separated by white space:
		(1) a letter in the query sequence alphabet;
		(2) a letter in the subject sequence alphabet;
		(3) the new pairwise score to be assigned to the alignment
		of these two letters.
		If the query (subject) letter is specified
		as the special word <TT>any</TT>,
		the altered score will be assigned
		to the entire column (row) of the scoring matrix.
		If the indicated score is the special word <TT>min</TT> (<TT>max</TT>),
		the new assigned score will be the minimum (maximum) score observed
		in the matrix.
		If the score is given as <TT>na</TT>,
		the alignment of the indicated letters will be <I>not allowed</I>,
		effectively assigning to them an infinite negative score.
		Multiple <CODE>altscore</CODE> options can be specified
		on a given command line.
		As an example of the option's use, to assign an alignment score
		of zero (0) to the presence of a stop codon
		in either the query or database sequence,
		these two specifications can be used together:
		<CODE>altscore="* any 0" altscore="any * 0"</CODE>.
		<BR>
		<I>See also:</I> <CODE><A HREF="#matrix">matrix</A></CODE>,
		<CODE><A HREF="#m">M</A></CODE> and <CODE><A HREF="#n">N</A></CODE>.
		</TD>
	</TR>
	<TR>
		<TD>
		&nbsp;&nbsp;<CODE>B=</CODE>&lt;<I>b</I>&gt;</TD>
	<TD><A NAME="b"></A>
<!-- PARM="B" -->
	set the maximum number of database sequences for which <I>any</I> alignments will be reported to <I>b</I>.
	The default limit is 250.
	The maximum number of alignments that may be saved and reported per
	database sequence is governed by other parameters.
		<BR>
		<I>See also:</I> <CODE><A HREF="#v">V</A></CODE>,
		<CODE><A HREF="#hspmax">hspmax</A></CODE>,
		<CODE><A HREF="#gspmax">gspmax</A></CODE>,
		<CODE><A HREF="#spoutmax">spoutmax</A></CODE> and
		<CODE><A HREF="#noseqs">noseqs</A></CODE>.
	</TD>
	</TR>
	<TR>
		<TD>
		&nbsp;&nbsp;<CODE>bottom</CODE></TD>
		<TD><A NAME="bottom"></A>
<!-- PARM="bottom" -->
		used to restrict the search of a nucleotide sequence
		to the bottom (-) strand.
		In the <TT>TBLASTX</TT> search mode, where both query and subject
		are nucleotide sequences, the <CODE>bottom</CODE> option only affects
		the query sequence.
		<BR>
		<I>See also:</I> <CODE><A HREF="#top">top</A></CODE>,
		<CODE><A HREF="#dbtop">dbtop</A></CODE>,
		<CODE><A HREF="#dbbottom">dbbottom</A></CODE> and
		<CODE><A HREF="#qframe">qframe</A></CODE>.
		</TD>
	</TR>
	<TR>
		<TD>
		&nbsp;&nbsp;<CODE>C=</CODE>&lt;<I>gcid</I>&gt;</TD>
		<TD><A NAME="c"></A>
<!-- PARM="C" -->
		use the indicated genetic code to translate the query
		sequence in the <TT>BLASTX</TT> and <TT>TBLASTX</TT> search modes.
		<I>gcid</I> is a numerical identifier for the desired code.
		A list of the genetic codes and their
		identifiers is displayed if <CODE>C=list</CODE> is specified
		on an otherwise syntactically correct command line.
		(Example:&nbsp;&nbsp;<CODE>blastx&nbsp;foo&nbsp;foo&nbsp;c=list</CODE>).
		In the <TT>TBLASTN</TT> search mode, the <CODE>C</CODE> parameter can be
		substituted for the <CODE>dbgcode</CODE> parameter.
<PRE>
The available genetic codes are:
   1. Standard*
   2. Vertebrate Mitochondrial
   3. Yeast Mitochondrial
   4. Mold Mitochondrial; Protozoan Mitochondrial; Coelenterate Mitochondrial;
        Mycoplasma; Spiroplasma
   5. Invertebrate Mitochondrial
   6. Ciliate Nuclear; Dasycladacean Nuclear; Hexamita Nuclear
   9. Echinoderm Mitochondrial
  10. Euplotid Nuclear
  11. Bacterial and Plant Plastid
  12. Alternative Yeast Nuclear
  13. Ascidian Mitochondrial
  14. Flatworm Mitochondrial
  15. Blepharisma Macronuclear
  16. Chlorophycean Mitochondrial
  21. Trematode Mitochondrial
  22. Scenedesmus obliquus mitochondrial
  23. Thraustochytrium mitochondrial
  1001. Codon2004
*The default genetic code (1).
Specify the desired genetic code by its number.
</PRE>
	The <CODE>Codon2004</CODE> code provides preliminary support
	for a draft alphabet for working precisely
	with each of the 64 possible codons,
	rather than mapping the codons to the usual 20 common amino acids.
	Scoring matrix files to use the Codon2004 alphabet
	with a translated query sequence in <TT>BLASTX</TT> should be
	placed in a subdirectory named <CODE>ca</CODE>,
	located parallel to the usual <CODE>aa</CODE> and <CODE>nt</CODE>
	subdirectories of the <CODE>matrix</CODE> directory.
	For use in <TT>TBLASTN</TT> searches, the scoring matrix should reside
	in an <CODE>ac</CODE> subdirectory;
	and for <TT>TBLASTX</TT> searches, the subdirectory should be <CODE>cc</CODE>.
	(Notice the use of the letter &ldquo;c&rdquo; for the codon alphabet,
	the letter &ldquo;a&rdquo; for the amino acid alphabet,
	and the query-subject ordering of the two letters to create
	the subdirectory name).
	For &ldquo;codon-ized&rdquo; scoring matrices derived from the BLOCKS database
	and appropriate for use &ldquo;as is&rdquo; with <TT>TBLASTX</TT>,
	please go
	<A HREF="http://dux.umbi.umd.edu/codon" TARGET="_blank">here</A>.
	For more information about the Codon2004 alphabet, please see
	<A HREF="http://dux.umbi.umd.edu/codon/mnemonic.html" TARGET="_blank">this</A>.
		<BR>
		<I>See also:</I> <CODE><A HREF="#dbgcode">dbgcode</A></CODE>.
		</TD>
	</TR>
	<TR>
		<TD>
		&nbsp;&nbsp;<CODE>cdb</CODE></TD>
		<TD><A NAME="cdb"></A>
<!-- PARM="cdb" -->
		force nucleotide sequence databases to be searched in their compressed form.
		This option is only effective in the <TT>BLASTN</TT> search mode for word lengths &ge; 7.
	<I>Users should generally avoid specifying this option themselves,
	letting the software decide when to employ this search strategy.</I>
		<BR>
		<I>See also:</I> <CODE><A HREF="#ucdb">ucdb</A></CODE>.
		</TD>
	</TR>
		<TR>
		<TD>
		&nbsp;&nbsp;<CODE>compat1.3</CODE></TD>
		<TD><A NAME="compat1.3"></A>
<!-- PARM="compat1.3" -->
		perform a BLAST version 1.3-style search (no gaps and significance estimated using Poisson statistics),
		but with bug fixes, performance enhancements and new options available.
		<BR>
		<I>See also:</I>  <CODE><A HREF="#compat1.4">compat1.4</A></CODE>.
		</TD>
	</TR>
	<TR>
		<TD>
		&nbsp;&nbsp;<CODE>compat1.4</CODE></TD>
		<TD><A NAME="compat1.4"></A>
<!-- PARM="compat1.4" -->
		perform a BLAST version 1.4-style search (no gaps in the alignments),
		but with bug fixes, performance enhancements and new options available.
		<BR>
		<I>See also:</I>  <CODE><A HREF="#compat1.3">compat1.3</A></CODE>.
		</TD>
	</TR>
	<TR>
		<TD>
		&nbsp;&nbsp;<CODE>consistency</CODE></TD>
		<TD><A NAME="consistency"></A>
<!-- PARM="consistency" -->
		turn off the determination of &ldquo;consistent&rdquo; sets
		of HSPs, effectively lumping all HSPs found for a given database
		sequence into one set.
		Use of this option also disables a combinatorial adjustment that is otherwise made
		to the Sum and Poisson statistics to account for the consistent
		arrangement of the HSPs out of all possible relative arrangements.
		This option has no effect if Sum or Poisson statistics are not being used.
		</TD>
	</TR>
	<TR>
		<TD>
		&nbsp;&nbsp;<CODE>cpus=</CODE>&lt;<I>n</I>&gt;</TD>
		<TD><A NAME="cpus"></A>
<!-- PARM="cpus" -->
		request that <I>n</I> processors or threads be employed for the search.
		The default behavior is to employ as many threads as there are
		processors in the computer system
		(to a maximum of 4 threads for <TT>BLASTN</TT> searches).
		This default may be altered by setting a specific value for <CODE>cpus</CODE>
		in a system-wide file named <CODE>/etc/sysblast</CODE>;
		see the <CODE>sysblast.sample</CODE> example file included in
		WU&nbsp;BLAST&nbsp;2.0 software distributions for further information.
		<I><TT>NOTE</TT>:</I>
		Memory consumption increases linearly with the number of threads;
		the actual number of threads employed may be automatically reduced
		by the software if memory resources are seen to be limiting.
		</TD>
	</TR>
	<TR>
		<TD>
		&nbsp;&nbsp;<CODE>ctxfactor=</CODE>&lt;<I>c</I>&gt;</TD>
		<TD><A NAME="ctxfactor"></A>
<!-- PARM="ctxfactor" -->
		set the &ldquo;context factor&rdquo; that is
		used as a Bonferroni correction in the statistics to <I>c</I>,
		to account for the number of contexts searched.
		Each distinct reading frame-to-reading frame or strand-to-strand combination
		between query and subject sequences constitutes one &ldquo;context&rdquo;.
		Thus, one context exists in a <TT>BLASTP</TT> search,
		as many as two contexts (because of the two distinct strand combinations) exist in a <TT>BLASTN</TT> search,
		up to 6 contexts (one for each reading frame) exist in a <TT>BLASTX</TT> or <TT>TBLASTN</TT> search,
		and up to 6x6 = 36 contexts exist in a <TT>TBLASTX</TT> search.
		The maximum default value for <CODE>ctxfactor</CODE> then is 1 for <TT>BLASTP</TT>,
		2 for <TT>BLASTN</TT>, 6 for <TT>BLASTX</TT> and <TT>TBLASTN</TT>,
		and 36 for <TT>TBLASTX</TT>.
		Restricting a search to a single strand of the query and/or database
		reduces the number of contexts accordingly for that search.
		More accurately, however,
		the contribution of any given context to the default value
		for <CODE>ctxfactor</CODE> is the fraction of residues in the query
		(or reading frame of the query) that are unambiguous (up to a maximum value of 1.0).
		(<I>N.B.</I> this fraction is computed <I>after</I> any optional filtering
		has been applied to the query).
		The default <CODE>ctxfactor</CODE> is then merely the sum of these fractions for every context involved in the search.
		<BR>
		<I>The software should normally be allowed to set the value of this
		parameter itself, unless the user has a compelling reason to change it.
		</I>
		One rationale for explicitly setting a value for <CODE>ctxfactor</CODE>
		might be to ensure a constant value is used in the statistics
		across multiple searches,
		where the results from the searches need to be examined
		and compared for their statistical significance on an common basis.
		</TD>
	</TR>
	<TR>
		<TD>
		&nbsp;&nbsp;<CODE>dbbottom</CODE></TD>
		<TD><A NAME="dbbottom"></A>
<!-- PARM="dbbottom" -->
		used to restrict the search to the bottom (-) strand of all database sequences.
		<BR>
		<I>See also:</I> <CODE><A HREF="#dbtop">dbtop</A></CODE>,
		<CODE><A HREF="#top">top</A></CODE>,
		<CODE><A HREF="#bottom">bottom</A></CODE> and
		<CODE><A HREF="#qframe">qframe</A></CODE>.
		</TD>
	</TR>
	<TR>
		<TD>
		&nbsp;&nbsp;<CODE>dbchunks=</CODE>&lt;<I>nchunks</I>&gt;</TD>
		<TD><A NAME="dbchunks"></A>
<!-- PARM="dbchunks" -->
		establishes the granularity of the database, as it is divided into
		slices for assignment to individual threads,
		to make more efficient use of all CPUs when multiple CPUs
		are employed for a given search.
		Higher values are appropriate when the database contains relatively
		few sequences and/or when the sequences vary greatly in length,
		composition or content (<I>e.g.,</I> genomic contigs).
		Lower values are appropriate when the database contains many
		sequences of comparable length
		(<I>e.g.,</I> the EST division of GenBank).
		The minimum assignable value is the number of threads employed,
		but this setting is ill-advised;
		the optimal value for any given search type is likely to be
		a large multiple of the number of threads employed
		(although it need not be an exact multiple).
		When searching mammalian genomic contigs, a good value may be 1000.
		The default value is 500.
		<BR>
		<I>Users generally need not be concerned with this parameter.</I>
		</TD>
	</TR>
	<TR>
		<TD>
		&nbsp;&nbsp;<CODE>dbgcode=</CODE>&lt;<I>gcid</I>&gt;</TD>
		<TD><A NAME="dbgcode"></A>
<!-- PARM="dbgcode" -->
		use the indicated genetic code to translate database
		sequences in the <TT>TBLASTN</TT> and <TT>TBLASTX</TT> search modes.
		<I>gcid</I> is a numerical identifier for the desired code.
		A list of the genetic codes and their
		identifers is displayed if <CODE>dbgcode=list</CODE> is specified
		on an otherwise syntactically correct command line.
		(Example:&nbsp;&nbsp;<CODE>tblastn&nbsp;foo&nbsp;foo&nbsp;dbgcode=list</CODE>).
		<BR>
		<I>See also:</I> <CODE><A HREF="#c">C</A></CODE>.
		</TD>
	</TR>
	<TR>
		<TD>
		&nbsp;&nbsp;<CODE>dbrecmax=</CODE>&lt;<I>last_record</I>&gt;</TD>
		<TD><A NAME="dbrecmax"></A>
<!-- PARM="dbrecmax" -->
		search the database until <I>last_record</I>,
		where database records are numbered starting with 1.
		By default, databases are searched completely.
		If <I>last_record</I> is greater than the actual
		number of records in the database, the database is simply
		searched until its end.
		It is an error for the requested <I>last_record</I> to be
		less than the first record requested to be searched in the database.
		Records in virtual databases are numbered with respect to the
		entire virtual database.
		<BR>
		<I>See also:</I> <CODE><A HREF="#dbrecmin">dbrecmin</A></CODE>.
		</TD>
	</TR>
	<TR>
		<TD>
		&nbsp;&nbsp;<CODE>dbrecmin=</CODE>&lt;<I>first_record</I>&gt;</TD>
		<TD><A NAME="dbrecmin"></A>
<!-- PARM="dbrecmin" -->
		search the database beginning at <I>first_record</I>,
		where database records are numbered starting with 1.
		By default, databases are searched completely.
		It is an error for the requested <I>first_record</I> to be
		greater than the last record requested to be searched in the database
		(re: the <CODE>dbrecmax</CODE> parameter)
		or to point beyond the end of the database.
		Records in virtual databases are numbered with respect to the
		entire virtual database.
		<BR>
		<I>See also:</I> <CODE><A HREF="#dbrecmax">dbrecmax</A></CODE>.
		</TD>
	</TR>
	<TR>
		<TD>
		&nbsp;&nbsp;<CODE>dbslice=</CODE><I>m</I>/<I>n</I>
		<BR>
		&nbsp;&nbsp;<CODE>dbslice=</CODE><I>a-b</I>/<I>n</I><BR>
		</TD>
		<TD><A NAME="dbslice"></A>
<!-- PARM="dbslice" -->
		at run time, for expressions of the form <I>m/n</I>,
		logically divide the database into <I>n</I> equivalent-sized slices and search only the <I>m</I><sup>th</sup> slice, where 1&nbsp;&le;&nbsp;<I>m</I>&nbsp;&le;&nbsp;<I>n</I>&nbsp;&le;&nbsp;100000.
		Alternatively, for expressions of the form <I>a-b/n</I>,
		search slices <I>a</I> through <I>b</I> (inclusive),
		where 1&nbsp;&le;&nbsp;<I>a</I>&nbsp;&le;&nbsp;<I>b</I>&nbsp;&le;&nbsp;<I>n</I>.
		Slice size is determined solely by the number of sequence <I>records</I> contained within and is not a function of sequence length.
		This can produce significant disparities in the workload
		associated with different slices, which may be alleviated
		by randomizing the order of sequences in the database before
		formatting for BLAST.
		In distributed computing environments,
		when the same, large database is to be searched repeatedly,
		overall throughput will likely benefit from consistently assigning
		the same slice(s) to the same client nodes for each search;
		improved efficiency results from the file caching activity
		that is typically performed by operating systems
		when the database files are first read from disk or over a network.
		Logically breaking the database into slices at run time means that
		each client node need only have sufficient unused memory
		as to be able to cache its assigned slice(s),
		not the entire database, and that the database
		need not be physically divided and reformatted into many
		smaller sub-databases whenever the number of available client nodes
		changes.
		</TD>
	</TR>
	<TR>
		<TD>
		&nbsp;&nbsp;<CODE>dbtop</CODE></TD>
		<TD><A NAME="dbtop"></A>
<!-- PARM="dbtop" -->
		used to restrict the search to the top (+) strand of all database sequences.
		<BR>
		<I>See also:</I> <CODE><A HREF="#dbbottom">dbbottom</A></CODE>,
		<CODE><A HREF="#top">top</A></CODE>,
		<CODE><A HREF="#bottom">bottom</A></CODE> and
		<CODE><A HREF="#qframe">qframe</A></CODE>.
		</TD>
	</TR>
	<TR>
		<TD>
		&nbsp;&nbsp;<CODE>E=</CODE>&lt;<I>e</I>&gt;</TD>
	<TD><A NAME="e"></A>
<!-- PARM="E" -->
	set the expectation threshold for reporting database hits to <I>e</I>.
	A database sequence will only be reported if an ascribed <I>E</I>-value
	for at least one of its alignments (or groups of alignments)
	is &le;&nbsp;<CODE>E</CODE>.
	Lower <I>E</I>-values are more significant (less likely to occur by chance).
	The default threshold is <CODE>E=10</CODE>,
	such that if the search algorithm exhibited 100% sensitivity
	and the statistics applied perfectly to the sequences being studied,
	results involving 10 database sequences would be reported merely by chance.
		<BR>
		<I>See also:</I> <CODE><A HREF="#s">S</A></CODE>.
	</TD>
	</TR>
	<TR>
		<TD>
		&nbsp;&nbsp;<CODE>E2=</CODE>&lt;<I>e</I>&gt;</TD>
	<TD><A NAME="e2"></A>
<!-- PARM="E2" -->
	set the expectation threshold for saving ungapped HSPs to <I>e</I>.
	In the initial, ungapped alignment phase of a search,
	individual HSPs will only be saved for further use
	if their score is &ge;&nbsp;<CODE>S2</CODE>,
	where the default value of <CODE>S2</CODE> is computed from <CODE>E2</CODE>.
	The default value for <CODE>E2</CODE> varies between <TT>BLAST</TT> search modes;
	the resultant value for <CODE>S2</CODE> will depend on the scoring system, as well.
	If both <CODE>E2</CODE> and <CODE>S2</CODE> are specified on the command line,
	the one corresponding to the more restrictive (higher) score threshold
	will be used.
		<BR>
		<I>See also:</I>
		<CODE><A HREF="#gape2">gapE2</A></CODE>,
		<CODE><A HREF="#s2">S2</A></CODE> and
		<CODE><A HREF="#gaps2">gapS2</A></CODE>.
	</TD>
	</TR>
	<TR>
		<TD>
		&nbsp;&nbsp;<CODE>echofilter</CODE></TD>
		<TD><A NAME="echofilter"></A>
<!-- PARM="echofilter" -->
		display the query sequence in the BLAST report, after all hard masks have been applied.
		<BR>
		<I>See also:</I> <CODE><A HREF="#filter">filter</A></CODE> and <CODE><A HREF="#lcfilter">lcfilter</A></CODE>.
		</TD>
	</TR>
	<TR>
		<TD>
		&nbsp;&nbsp;<CODE>endgetenv</CODE></TD>
		<TD><A NAME="endgetenv"></A>
<!-- PARM="endgetenv" -->
		ignore any subsequent <CODE><A HREF="#getenv">getenv</A></CODE> options found on the command line during left-to-right parsing.
		<BR>
		<I>See also:</I> <CODE><A HREF="#endputenv">endputenv</A></CODE>,
		<CODE><A HREF="#getenv">getenv</A></CODE> and <CODE><A HREF="#putenv">putenv</A></CODE>.
		</TD>
	</TR>
	<TR>
		<TD>
		&nbsp;&nbsp;<CODE>endputenv</CODE></TD>
		<TD><A NAME="endputenv"></A>
<!-- PARM="endputenv" -->
		for security in WWW server installations, where the command line may sometimes be left open to users,
		ignore any subsequent <CODE><A HREF="#putenv">putenv</A></CODE> options found on the command line during left-to-right parsing.
		<BR>
		<I>See also:</I> <CODE><A HREF="#endgetenv">endgetenv</A></CODE>,
		<CODE><A HREF="#getenv">getenv</A></CODE> and <CODE><A HREF="#putenv">putenv</A></CODE>.
		</TD>
	</TR>
	<TR>
		<TD>
		&nbsp;&nbsp;<CODE>errors</CODE></TD>
		<TD><A NAME="errors"></A>
<!-- PARM="errors" -->
		suppress all <CODE>ERROR</CODE> messages.  These messages should rarely, if ever, arise and indicate severe conditions (typically internal software bugs) that should be given immediate attention.  When they do arise, parsers may break.
		If any <CODE>ERROR</CODE>s arise with this option, the number <CODE>SUPPRESSED</CODE> will be reported
		at the end of the search.
		<BR>
		<I>See also:</I>
		<CODE><A HREF="#notes">notes</A></CODE>
		and
		<CODE><A HREF="#warnings">warnings</A></CODE>.
		</TD>
	</TR>
	<TR>
		<TD>
		&nbsp;&nbsp;<CODE>evalues</CODE></TD>
		<TD><A NAME="evalues"></A>
<!-- PARM="evalues" -->
		report <I>E</I>-values (expectations) instead of <I>P</I>-values (probabilities) in the initial one-line descriptions section of output.
		<BR>
		<I>See also:</I> <CODE><A HREF="#pvalues">pvalues</A></CODE>.
		</TD>
	</TR>
	<TR>
		<TD>
		&nbsp;&nbsp;<CODE>filter=</CODE>&lt;<I>filter</I>&gt;</TD>
		<TD><A NAME="filter"></A>
<!-- PARM="filter" -->
		&ldquo;hard mask&rdquo; the query sequence using the specified <I>filter</I>.
		The filter program may alter the sequence in composition but not in length.
		For protein-level searches (<TT>BLASTP</TT>, <TT>BLASTX</TT>, <TT>TBLASTN</TT> and <TT>TBLASTX</TT>), the supported filter programs include:
		<CODE>seg</CODE> and <CODE>xnu</CODE>.
		For nucleotide-level (<TT>BLASTN</TT>) searches, supported filter programs include:
		<CODE>dust</CODE> and <CODE>seg</CODE>.
		If multiple <CODE>filter</CODE> specifications are made on the command line, their results are logically OR-ed.
		<BR>
		<CODE>filter=none</CODE> causes any earlier specifications (to the left) on the command line to be ignored.
		<BR>
		<B>NOTE:  By default, no filtering is performed.</B>
		<BR>
		Arbitrary user-defined filter programs can be utilized,
		if their input and output are sequences in FASTA/Pearson format
		and if input/output are tied to stdin/stdout.
		<BR>
		The location of filter programs is governed by the <CODE>BLASTFILTER</CODE>
		environment variable, which can be set to a colon-delimited list of directories that the <TT>BLAST</TT> programs will successively examine to find filters.
		<BR>
		<I>See also:</I>
		<CODE><A HREF="#wordmask">wordmask</A></CODE>,
		<CODE><A HREF="#lcfilter">lcfilter</A></CODE>,
		<CODE><A HREF="#lcmask">lcmask</A></CODE> and
		<CODE><A HREF="#echofilter">echofilter</A></CODE>.
		</TD>
	</TR>
	<TR>
		<TD>
		&nbsp;&nbsp;<CODE>gapall</CODE></TD>
		<TD><A NAME="gapall"></A>
<!-- PARM="gapall" -->
		effectively generate a gapped alignment for every ungapped HSP found (up to <CODE><A HREF="#hspmax">hspmax</A></CODE>).  This is the default behavior.
		<BR>
		<I>See also:</I> <CODE><A HREF="#gape">gapE</A></CODE>.
		</TD>
	</TR>
	<TR>
		<TD>
		&nbsp;&nbsp;<CODE>gapdecayrate=</CODE>&lt;<I>r</I>&gt;</TD>
	<TD><A NAME="gapdecayrate"></A>
<!-- PARM="gapdecayrate" -->
	define <I>r</I> to be the common ratio of the terms in a geometric progression
	used in altering probabilities as a function of the number of Poisson
	events involved (typically the number of &ldquo;consistent&rdquo; HSPs in a set),
	according to a method suggested by Phil Green.
	An initial Poisson probability for <I>n</I> HSPs is weighted
	by the quantity <I>T<sub>n</sub></I>, which is itself
	the reciprocal of the <I>n</I><sup>th</sup> term in the progression
	<I>t<sub>n</sub></I>&nbsp;=&nbsp;(1-<I>r</I>)<I>r</I><sup><I>n</I>-1</sup>.
	The default value for <I>r</I> is 0.5, such that
	the default weights are successively <I>T</I><sub>1</sub>=2,
	<I>T</I><sub>2</sub>=4,
	<I>T</I><sub>3</sub>=8,
	<I>T</I><sub>4</sub>=16, and so on.
	These weights provide a conservative Bonferroni correction to the probabilities,
	in case multiple trials are performed in determining the set
	of HSPs yielding the lowest <I>P</I>-value for a given database sequence.
	That the geometric progression contains an infinite number of terms
	allows this correction method to satisfy the need for any number
	of tests, when this number is unknown prior to the search.
	<BR>
	The value for <CODE>gapdecayrate</CODE> affects the statistics
	when the default Sum statistics or optional Poisson statistics are used,
	but not when multiple HSP statistics have been turned off
	with the <CODE><A HREF="#kap">kap</A></CODE> option.
	</TD>
	</TR>
	<TR>
		<TD>
		&nbsp;&nbsp;<CODE>gapE=</CODE>&lt;<I>gapE</I>&gt;</TD>
		<TD><A NAME="gape"></A>
<!-- PARM="gapE" -->
		generate gapped alignments for all HSPs between sequences whose expected frequency of chance occurrence is &le;&nbsp;<I>gapE</I>.
		Default value is <CODE>gapE=</CODE><I>infinity</I> &mdash; <I>i.e.,</I> <CODE>gapall</CODE> is in effect.
		<BR>
		<I>See also:</I>
		<CODE><A HREF="#gapall">gapall</A></CODE>,
		</TD>
	</TR>
	<TR>
		<TD>
		&nbsp;&nbsp;<CODE>gapE2=</CODE>&lt;<I>e</I>&gt;</TD>
	<TD><A NAME="gape2"></A>
<!-- PARM="gapE2" -->
	set the <I>E</I>-value for saving gapped HSPs to <I>e</I>.
	In the secondary, gapped alignment phase of a search,
	individual gapped HSPs will only be saved for further use
	if their score is &ge;&nbsp;<CODE>gapS2</CODE>,
	where the default <CODE>gapS2</CODE> is computed from <CODE>gapE2</CODE>.
	The default value for <I>gapE2</I> varies between <TT>BLAST</TT> search modes;
	the resultant <CODE>gapS2</CODE> will depend on the scoring system, as well.
	If both <CODE>gapE2</CODE> and <CODE>gapS2</CODE> are specified on the command line,
	the one corresponding to the more restrictive (higher) score threshold
	will be used.
		<BR>
		<I>See also:</I>
		<CODE><A HREF="#gaps2">gapS2</A></CODE>,
		<CODE><A HREF="#e2">E2</A></CODE> and
		<CODE><A HREF="#s2">S2</A></CODE>.
	</TD>
	</TR>
	<TR>
		<TD>
		&nbsp;&nbsp;<CODE>gapH=</CODE>&lt;<I>h</I>&gt;</TD>
		<TD><A NAME="gaph"></A>
<!-- PARM="gapH" -->
		set the value of the relative entropy, <I>H</I>, used in evaluating the statistical significance of gapped alignment scores.
		<BR>
		<I>See also</I> <CODE><A HREF="#h">H</A></CODE>.
		</TD>
	</TR>
	<TR>
		<TD>
		&nbsp;&nbsp;<CODE>gapK=</CODE>&lt;<I>k</I>&gt;</TD>
		<TD><A NAME="gapk"></A>
<!-- PARM="gapK" -->
		set the value of the extreme value statistics <I>K</I> parameter
		(<A HREF="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=Retrieve&amp;db=pubmed&amp;dopt=Abstract&amp;list_uids=2315319">Karlin and Altschul, 1990</A>)
		used in evaluating the significance of gapped alignment scores.
		Useful when precomputed values are unavailable in the internal tables for the chosen scoring matrix and gap penalty combination.
		<BR>
		<I>See also</I> <CODE><A HREF="#k">K</A></CODE>.
		</TD>
	</TR>
	<TR>
		<TD>
		&nbsp;&nbsp;<CODE>gapL=</CODE>&lt;<I>lambda</I>&gt;</TD>
		<TD><A NAME="gapl"></A>
<!-- PARM="gapL" -->
		use <I>lambda</I> for the value of the
		<I>&lambda;</I> parameter in the extreme value statistics
		used to evaluate the significance of gapped alignment scores
		(<A HREF="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=Retrieve&amp;db=pubmed&amp;dopt=Abstract&amp;list_uids=8743700">Altschul and Gish, 1996</A>).
		Useful when precomputed values are unavailable in the internal tables for the chosen scoring matrix and gap penalty combination.
		<BR>
		<I>See also:</I> <CODE><A HREF="#l">L</A></CODE>.
		</TD>
	</TR>
	<TR>
		<TD>
		&nbsp;&nbsp;<CODE>gaps</CODE></TD>
		<TD><A NAME="gaps"></A>
<!-- PARM="gaps" -->
		produce gapped alignments (the default behavior),
		negating the effect of any previously specified <CODE>nogaps</CODE> option.
		<BR>
		<I>See also:</I> <CODE><A HREF="#nogaps">nogaps</A></CODE> and <CODE><A HREF="#gapall">gapall</A></CODE>.
		</TD>
	</TR>
	<TR>
		<TD>
		&nbsp;&nbsp;<CODE>gapS2=</CODE>&lt;<I>s</I>&gt;</TD>
	<TD><A NAME="gaps2"></A>
<!-- PARM="gapS2" -->
	set the score threshold for saving gapped HSPs to <I>s</I>.
	In the secondary, gapped alignment phase of a search,
	individual gapped HSPs will only be saved for further use
	if their score is &ge;<CODE>gapS2</CODE>.
	The default score threshold is computed from <CODE>gapE2</CODE>
	and will depend on the scoring system.
	If both <CODE>E2</CODE> and <CODE>S2</CODE> are specified on the command line,
	the one corresponding to the more restrictive (higher) score threshold
	will be used.
		<BR>
		<I>See also:</I>
		<CODE><A HREF="#gape2">gapE2</A></CODE>,
		<CODE><A HREF="#e2">E2</A></CODE> and
		<CODE><A HREF="#s2">S2</A></CODE>.
	</TD>
	</TR>
	<TR>
		<TD>
		&nbsp;&nbsp;<CODE>gapW=</CODE>&lt;<I>gapW</I>&gt;</TD>
		<TD><A NAME="gapw"></A>
<!-- PARM="gapW" -->
		set the window width (or band width) within which gapped alignments are computed by dynamic programming (default is <CODE>gapW=32</CODE> for protein comparisons, <CODE>gapW=16</CODE> for <TT>BLASTN</TT>).  Note:&nbsp;<CODE>gapW</CODE> is the full bandwidth, not the half-width.</TD>
	</TR>
	<TR>
		<TD>
		&nbsp;&nbsp;<CODE>gapX=</CODE>&lt;<I>x</I>&gt;</TD>
	<TD><A NAME="gapx"></A>
<!-- PARM="gapX" -->
	set the drop-off score for gapped alignment extensions to <I>x</I>.
	Gapped extension of ungapped HSPs found
	between query and subject sequences
	continues until the cumulative alignment score deteriorates
	from the maximum value seen thusfar by a quantity <I>gapX</I> or more.
	The default value for <CODE>gapX</CODE> is the score associated with 10 bits
	of significance (2<sup>-10</sup> &lt; 10<sup>-3</sup> probability) for protein-level searches or 20 bits
	of significance (2<sup>-20</sup> &lt; 10<sup>-6</sup> probability)
	for nucleotide-level (<TT>BLASTN</TT>) searches.
	Higher values for <CODE>gapX</CODE> will increase sensitivity at the expense
	of run time.
		<BR>
		<I>See also:</I> <CODE><A HREF="#x">X</A></CODE> and
		<CODE><A HREF="#gapw">gapW</A></CODE>.
	</TD>
	</TR>
	<TR>
		<TD>
		&nbsp;&nbsp;<CODE>getenv=</CODE>"<I><FONT SIZE=-1>NAME</FONT></I>"</TD>
		<TD><A NAME="getenv"></A>
<!-- PARM="getenv" -->
		display the value of the environment variable named <I>NAME</I>.  This may be useful for verifying that the settings of environment variables
		on a web server or in an analysis pipeline have been propagated all the way to the BLAST search program.
		<BR>
		<I>See also:</I> <CODE><A HREF="#endgetenv">endgetenv</A></CODE>,
		<CODE><A HREF="#putenv">putenv</A></CODE> and <CODE><A HREF="#endputenv">endputenv</A></CODE>.
		</TD>
	</TR>
	<TR>
		<TD>
		&nbsp;&nbsp;<CODE>gi</CODE></TD>
		<TD><A NAME="gi"></A>
<!-- PARM="gi" -->
		report NCBI &ldquo;gi&rdquo; (GenInfo) identifiers for sequences,
		when present in sequence definition lines.
		Normally these identifiers are suppressed from output,
		but they represent one of the best, stable identifiers available
		for the GenBank/EMBL/DDBJ databases
		(with <CODE>ACCESSION.VERSION</CODE> being the other stable identifier).
		</TD>
	</TR>
	<TR>
		<TD>
		&nbsp;&nbsp;<CODE>globalexit</CODE></TD>
		<TD><A NAME="globalexit"></A>
<!-- PARM="globalexit" -->
		when processing a file containing multiple query sequences,
		if any of them encounters a <CODE>FATAL</CODE> error,
		then after all queries have been processed,
		append the line "<CODE>EXIT CODE 12</CODE>" to the output and provide a testable
		exit status 12;
		if the exit status is 0 or if the last line of output is not "<CODE>EXIT STATUS 12</CODE>", then it can be assumed that all queries succeeded.
		To determine whether all queries succeeded without this option,
		the output would need to be scanned for instances of <CODE>EXIT CODE</CODE>
		with a non-zero argument.
		With the <CODE>globalexit</CODE> option, scanning of the output
		is only necessary when one wishes to identify the specific query (or queries)
		that failed and what the individual reason codes were.
		<BR>
		<I>See also:</I> <CODE><A HREF="#haltonfatal">haltonfatal</A></CODE>.
		</TD>
	</TR>
	<TR>
		<TD>
		&nbsp;&nbsp;<CODE>golfraction=</CODE>&lt;<I>g</I>&gt;</TD>
		<TD><A NAME="golfraction"></A>
<!-- PARM="golfraction" -->
		maximum fractional length of overlap, <I>g</I>, of two <I>gapped</I> alignments for them to be considered independent and mutually &ldquo;consistent&rdquo; and their joint (Sum or Poisson) probability to be computed.  The default value is 0.125 (maximum 12.5% of the length from either end of either HSP).
		For any given pair of HSPs, the more restrictive of <CODE>golfraction</CODE>
		and <CODE>golmax</CODE> is used.
		To eliminate <CODE>golfraction</CODE> from consideration,
		set its value to 1,
		to indicate the acceptability of even a complete, 100% overlap.
		<BR>
		<I>See also:</I>
		<CODE><A HREF="#golmax">golmax</A></CODE>,
		<CODE><A HREF="#olfraction">olfraction</A></CODE>,
		and
		<CODE><A HREF="#olmax">olmax</A></CODE>.
		</TD>
	</TR>
	<TR>
		<TD>
		&nbsp;&nbsp;<CODE>golmax=</CODE>&lt;<I>len</I>&gt;</TD>
		<TD><A NAME="golmax"></A>
<!-- PARM="golmax" -->
		set the maximum permitted length of overlap (in residues), <I>len</I>, of two <I>gapped</I> alignments for their joint (Sum or Poisson) probability to be computed.  The default is unlimited length, with the maximum extent of overlap being governed only by the <CODE><A HREF="#golfraction">golfraction</A></CODE> parameter.
		<BR>
		<I>See also:</I>
		<CODE><A HREF="#golfraction">golfraction</A></CODE>,
		<CODE><A HREF="#olfraction">olfraction</A></CODE>,
		and
		<CODE><A HREF="#olmax">olmax</A></CODE>.
		</TD>
	</TR>
	<TR>
		<TD>
		&nbsp;&nbsp;<CODE>gspmax=</CODE>&lt;<I>gspmax</I>&gt;</TD>
		<TD><A NAME="gspmax"></A>
<!-- PARM="gspmax" -->
		establish <I>gspmax</I> as the maximum number of GSPs (gapped HSPs)
		to report per subject sequence or pairwise sequence comparison.
		If more than <CODE>gspmax</CODE> GSPs are found,
		only the best-scoring GSPs are retained for subsequent processing and reporting.
		The setting of <CODE>gspmax</CODE> will have no effect
		if the <CODE><A HREF="#nogaps">nogaps</A></CODE> option is specified or
		if the setting of <CODE><A HREF="#hspmax">hspmax</A></CODE> is more restrictive.
		<BR>
		The default value for <CODE>gspmax</CODE> is 0, which implies no limit.
		<BR>
		<I>See also:</I> <CODE>hspmax</CODE>, <CODE><A HREF="#spoutmax">spoutmax</A></CODE>.
		<BR>
		<I><TT>NOTE</TT>:</I> the <CODE><A HREF="#b">B</A></CODE> and <CODE><A HREF="#v">V</A></CODE> options limit the number
		of subject sequences for which any results whatsoever are reported,
		regardless of the number of HSPs or GSPs found.
		</TD>
	</TR>
	<TR>
		<TD>
		&nbsp;&nbsp;<CODE>H=</CODE>&lt;<I>h</I>&gt;</TD>
		<TD><A NAME="h"></A>
<!-- PARM="H" -->
		use <I>h</I> for the value of the relative entropy, <I>H</I>,
		when computing the statistics of ungapped alignments.
		<BR>
		<I><TT>NOTE</TT>:</I> In BLAST 1.4 and earlier, the <CODE>H</CODE> option was used to invoke the display of a histogram of search results; this functionality is no longer supported.)
		<BR>
		<I>See also:</I> <CODE><A HREF="#gaph">gapH</A></CODE>.
		</TD>
	</TR>
	<TR>
		<TD>
		&nbsp;&nbsp;<CODE>haltonfatal</CODE></TD>
		<TD><A NAME="haltonfatal"></A>
<!-- PARM="haltonfatal" -->
		when processing a file containing multiple query sequences, use this option to
		halt further processing at the first occurrence of a <CODE>FATAL</CODE> error.
		Processing will otherwise resume with the next query sequence
		when a <CODE>FATAL</CODE> error arises.
		<BR>
		<I>See also:</I> <CODE><A HREF="#globalexit">globalexit</A></CODE>.
		</TD>
	</TR>
	<TR>
		<TD>
		&nbsp;&nbsp;<CODE>hitdist=</CODE>&lt;<I>hitdist</I>&gt;</TD>
		<TD><A NAME="hitdist"></A>
<!-- PARM="hitdist" -->
		invoke a 2-hit BLAST algorithm similar to (but more sensitive and efficient than) that of
		<A HREF="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=Retrieve&amp;db=pubmed&amp;dopt=Abstract&amp;list_uids=9254694">
		Altschul <I>et al.</I> (1997)</A>,
		with the maximum distance between word hits along the same diagonal of &lt;<I>hitdist</I>&gt; residues, for seeding ungapped extensions.
		<A HREF="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=Retrieve&amp;db=pubmed&amp;dopt=Abstract&amp;list_uids=9254694">
		Altschul <I>et al.</I> (1997)</A>
		use the equivalent of <CODE>hitdist=40</CODE>
in the <TT>BLASTP</TT>, <TT>BLASTX</TT>, <TT>TBLASTN</TT> and <TT>TBLASTX</TT> search modes.
In WU&nbsp;<TT>BLASTN</TT>, setting <CODE>hitdist=</CODE><I>W</I> and <CODE>wink=</CODE><I>W</I>,
where <I>W</I> is the word length, is akin to using double-length words generated
on <I>W</I>-mer boundaries.
		<BR>
		<I><TT>NOTE</TT>:</I> In protein-level comparisons, for best sensitivity (or the best sensitivity for the amount of memory used),
		2-hit BLAST is not recommended.
		<BR>
		<I>See also:</I> <CODE><A HREF="#wink">wink</A></CODE>.
	</TD>
	</TR>
	<TR>
		<TD>
		&nbsp;&nbsp;<CODE>hspmax=</CODE>&lt;<I>hspmax</I>&gt;</TD>
		<TD><A NAME="hspmax"></A>
<!-- PARM="hspmax" -->
		establishes <I>hspmax</I> as the maximum number of ungapped HSPs
		that will be saved per subject sequence or pairwise sequence comparison.
		Saved HSPs are then fed to the gapped alignment phase of the program
		or are statistically evaluated
		if gapped alignments are not to be performed.
		If more than <CODE>hspmax</CODE> HSPs are found,
		only the best-scoring HSPs are retained for subsequent processing.
		<BR>
		The default value is 1000; a value of 0 signifies no limit.
		<BR>
		<I>See also:</I>
		<CODE><A HREF="#gspmax">gspmax</A></CODE> and
		<CODE><A HREF="#spoutmax">spoutmax</A></CODE>.
		<BR>
		<I><TT>NOTE</TT>:</I> This usage of <CODE>hspmax</CODE> is subtly,
		but importantly,
		different from the parameter's classical interpretation,
		wherein <I>all</I> ungapped HSPs that satisfied the S2 score threshold
		were saved; <CODE>hspmax</CODE> merely limited
		the number of HSPs (gapped or ungapped) that would be reported.
		The new interpretation was instituted to provide
		vastly improved speed on large problems,
		while imparting no effect on small problems
		and many medium-sized problems.
		The new behavior can help guard against horrendously slow searches
		resulting from an inadvertent omission of a low-complexity filter.
		Adverse effects on sensitivity may be obtained, however,
		if every HSP is sacred.
		To restore classical behavior, specify <CODE>hspmax=0</CODE>.
		As a compromise between sensitivity and speed, set a higher
		value than the default.
		<BR>
		<I><TT>NOTE</TT>:</I> the <CODE><A HREF="#b">B</A></CODE> and <CODE><A HREF="#v">V</A></CODE> options limit the number
		of database or subject sequences for which any results are reported,
		regardless of the number of HSPs or GSPs found.
		</TD>
	</TR>
	<TR>
		<TD>
		&nbsp;&nbsp;<CODE>hspsepQmax=</CODE>&lt;<I>d</I>&gt;</TD>
		<TD><A NAME="hspsepqmax"></A>
<!-- PARM="hspsepQmax" -->
		maximum allowed separation along the <I>query</I> sequence between two HSPs (gapped or ungapped) that will be clustered into a &ldquo;consistent&rdquo; set.
		Distance is measured here in units of residues <I>at the level of the actual sequence comparison</I>
		&mdash; <I>i.e.</I>, in nucleotides for <TT>BLASTN</TT> and in peptides (or codons) for all other search modes.
		This option is useful for improving the statistical power to discriminate clusters
		that have potential biological interest from random background clusters,
		when the query sequence is significantly longer than the features of interest.
		Without this restriction, HSPs may be linked that arise
		from very distant portions of the query sequence.
		Depending on the specific search performed,
		distant links may be desirable,
		but often a reasonable setting for this parameter might be
		the expected maximum length of an intron.
		A distance restriction not only avoids clustering HSPs that
		would be widely separated but improves the statistics of those
		HSPs that still can be clustered.
		</TD>
	</TR>
	<TR>
		<TD>
		&nbsp;&nbsp;<CODE>hspsepSmax=</CODE>&lt;<I>d</I>&gt;</TD>
		<TD><A NAME="hspsepsmax"></A>
<!-- PARM="hspsepSmax" -->
		maximum allowed separation along the <I>subject</I> (database) sequence between two HSPs (gapped or ungapped) that will be clustered into a consistent set.
		Distance is measured here in units of residues <I>at the level of the actual sequence comparison</I>
		&mdash; <I>i.e.</I>, in nucleotides for <TT>BLASTN</TT> and in peptides (or codons) for all other search modes.
		This option is useful for improving the statistical power to discriminate clusters
		that have potential biological interest from random background clusters,
		when the database contains sequences significantly longer than the features of interest.
		Without this restriction, HSPs may be linked that arise
		from very distant portions of a subject sequence.
		Depending on the specific search performed,
		distant links may be desirable,
		but often a reasonable setting for this parameter might be
		the expected maximum length of an intron.
		A distance restriction not only avoids clustering HSPs that
		would be widely separated but improves the statistics of those
		HSPs that still can be clustered.
		</TD>
	</TR>
	<TR>
		<TD>
		&nbsp;&nbsp;<CODE>K=</CODE>&lt;<I>k</I>&gt;</TD>
		<TD><A NAME="k"></A>
<!-- PARM="K" -->
		set the value for extreme value statistics <I>K</I> parameter
		(<A HREF="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=Retrieve&amp;db=pubmed&amp;dopt=Abstract&amp;list_uids=2315319">Karlin and Altschul, 1990</A>)
		used in computing the statistics of ungapped alignments.
		<BR>
		<I>See also:</I> <CODE><A HREF="#gapk">gapK</A></CODE>.
		</TD>
	</TR>
	<TR>
		<TD>
		&nbsp;&nbsp;<CODE>kap</CODE></TD>
		<TD><A NAME="kap"></A>
<!-- PARM="kap" -->
		use basic
		<A HREF="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=Retrieve&amp;db=pubmed&amp;dopt=Abstract&amp;list_uids=2315319">Karlin and Altschul (1990)</A>
		statistics on individual alignment scores (<I>i.e.,</I> do not evaluate the joint probability of multiple consistent HSP scores, such as with Poisson or
		<A HREF="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=Retrieve&amp;db=pubmed&amp;dopt=Abstract&amp;list_uids=8390686">Karlin and Altschul (1993)</A>
		&ldquo;Sum&rdquo; statistics);
		in order to be reported,
		each HSP must pass the significance test on its own;
		these basic statistics are an option in all search modes.
		</TD>
	</TR>
	<TR>
		<TD>
		&nbsp;&nbsp;<CODE>L=</CODE>&lt;<I>lambda</I>&gt;</TD>
		<TD><A NAME="l"></A>
<!-- PARM="L" -->
		use <I>lambda</I> for the value of the
		<I>&lambda;</I> parameter in the extreme value statistics
		(<A HREF="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=Retrieve&amp;db=pubmed&amp;dopt=Abstract&amp;list_uids=2315319">Karlin and Altschul, 1990</A>)
		used in computing the statistics of ungapped alignments.
		<BR>
		<I>See also:</I> <CODE><A HREF="#gapl">gapL</A></CODE>.
		</TD>
	</TR>
	<TR>
		<TD>
		&nbsp;&nbsp;<CODE>lcfilter</CODE></TD>
		<TD><A NAME="lcfilter"></A>
<!-- PARM="lcfilter" -->
		replace any lower case letters in the input query sequence
		with the appropriate ambiguity code for &ldquo;any&rdquo; residue
		(N for nucleotide sequences; X for protein sequences).
		<BR>
		<I>See also:</I>
		<CODE><A HREF="#lcmask">lcmask</A></CODE>,
		<CODE><A HREF="#filter">filter</A></CODE>,
		<CODE><A HREF="#wordmask">wordmask</A></CODE> and
		<CODE><A HREF="#echofilter">echofilter</A></CODE>.
		</TD>
	</TR>
	<TR>
		<TD>
		&nbsp;&nbsp;<CODE>lcmask</CODE></TD>
		<TD><A NAME="lcmask"></A>
<!-- PARM="lcmask" -->
		when generating the neighborhood word list for the query sequence,
		do not process any portions of the query that were represented
		in lower case letters in the input file.
		Lower case letters in the query sequence remain unchanged
		by this &ldquo;soft masking&rdquo; procedure and can therefore participate in alignments
		seeded by word hits that occur in flanking regions.
		<BR>
		<I>See also:</I>
		<CODE><A HREF="#lcfilter">lcfilter</A></CODE>,
		<CODE><A HREF="#wordmask">wordmask</A></CODE>,
		<CODE><A HREF="#filter">filter</A></CODE>,
		<CODE><A HREF="#maskextra">maskextra</A></CODE> and
		<CODE><A HREF="#echofilter">echofilter</A></CODE>.
		</TD>
	</TR>
	<TR>
		<TD>
		&nbsp;&nbsp;<CODE>links</CODE></TD>
		<TD><A NAME="links"></A>
<!-- PARM="links" -->
		report consistent link information for each alignment, indicating the set of &ldquo;consistent&rdquo; alignments used in joint statistical
		 significance calculations.
		 Links information appears on its own line for each HSP
		 and begins with the keyword <CODE>Links</CODE>.
		 Each HSP involving the query and a given subject sequence
		 is numbered from 1 to <I>n</I>,
		 where <I>n</I> is the total number of HSPs reported
		 for the pair of sequences.
		 When the <CODE>links</CODE> option is specified,
		 the current HSP number is enclosed in parenthese.
		 <P>
		 For example, the links information for an HSP might
		 look like the following,
		 where the HSP number 1 enclosed in parentheses indicates
		 that this information accompanied the first HSP reported for the given subject
		 sequence.
		 It is evident in this example that a total of at least 8 HSPs were reported
		 for the subject sequence (re: the 8 in the links list),
		 but only 3 consistent HSPs (numbers 8, 2 and 1, in that order)
		 were involved in obtaining the Sum statistics <I>P</I>-value of 0.15.
		 </P>
		 <PRE>
		 Score = 72 (30.4 bits), Expect = 0.16, Sum P(3) = 0.15
		 Identities = 41/174 (23%), Positives = 74/174 (42%)
		 Links = 8-2-(1)
		 </PRE>
		<I><TT>NOTE</TT>:</I> While all link lists describe sets of consistent HSPs,
		unless one of the <CODE>topcomboN</CODE>
		or <CODE>topcomboE</CODE> options is used,
		only the list reported for HSPs in the most significant set
		for each subject sequence is guaranteed to represent the
		precise set of HSPs for which the joint statistics were computed;
		all other link lists often do correctly describe the set of HSPs
		involved but may have one or more missing or extraneous HSPs.
		<BR>
		<I>See also:</I> <CODE><A HREF="#hspsepqmax">hspsepQmax</A></CODE>,
		<CODE><A HREF="#hspsepsmax">hspsepSmax</A></CODE>,
		<CODE><A HREF="#topcomboe">topcomboE</A></CODE> and
		<CODE><A HREF="#topcombon">topcomboN</A></CODE>.
		</TD>
	</TR>
	<TR>
		<TD>
		&nbsp;&nbsp;<CODE>M=</CODE>&lt;<I>m</I>&gt;</TD>
		<TD><A NAME="m"></A>
<!-- PARM="M" -->
		set the positive reward score for matching nucleotides in the <TT>BLASTN</TT>
		search mode to <I>m</I>, with default value +5.
	<BR>
	For compatibility with earlier versions of <TT>BLAST</TT>,
	in search modes other than <TT>BLASTN</TT>,
	the <CODE>M</CODE> option is synonymous with the
	<CODE>matrix</CODE> option.
	To provide a fully specified scoring matrix to <TT>BLASTN</TT>,
	the <CODE>matrix</CODE> option itself must be used.
		<BR>
		<I>See also:</I>
		<CODE><A HREF="#n">N</A></CODE>,
		<CODE><A HREF="#matrix">matrix</A></CODE> and
		<CODE><A HREF="#altscore">altscore</A></CODE>.
	</TD>
	</TR>
	<TR>
		<TD>
		&nbsp;&nbsp;<CODE>maskextra=</CODE>&lt;<I>extra</I>&gt;</TD>
		<TD><A NAME="maskextra"></A>
<!-- PARM="maskextra" -->
		soft mask for an additional <I>extra</I> letters
		to each side of regions that are soft masked by the
		<CODE>lcmask</CODE> and <CODE>wordmask</CODE> options.
		This reduces the incidence of high scoring alignments
		in low-complexity regions that would be
		initiated by spurious word hits
		in otherwise unmasked flanking regions.
		<BR>
		<I>See also:</I>
		<CODE><A HREF="#wordmask">wordmask</A></CODE>,
		<CODE><A HREF="#lcmask">lcmask</A></CODE> and
		<CODE><A HREF="#lcfilter">lcfilter</A></CODE>.
		</TD>
	 </TR>
	<TR>
		<TD>
		&nbsp;&nbsp;<CODE>matrix=</CODE>&lt;<I>name</I>&gt;</TD>
	<TD><A NAME="matrix"></A>
<!-- PARM="matrix" -->
	use the 2-dimensional matrix named <I>name</I> to score residue pairs
	in gapped and ungapped alignments.
	The default matrix for protein-level searches is BLOSUM62
	(<A HREF="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=Retrieve&amp;db=pubmed&amp;dopt=Abstract&amp;list_uids=1438297">Henikoff and Henikoff, 1992</A>).
	For <TT>BLASTN</TT> searches,
	the default scoring matrix is computed
	dynamically from a +5/-4 match/mismatch scoring system
	which can be altered using the <CODE>M</CODE> and <CODE>N</CODE> parameters.
	<TT>BLASTN</TT> can also use fully specified scoring matrices
	of the user's own design,
	by providing the name of the matrix with the <CODE>matrix</CODE> option.
	After unpacking the software, see the <CODE>matrix/nt</CODE> subdirectory
	for some examples of nucleotide scoring matrices.
	<BR>
	<I><TT>NOTE</TT>:</I> matrices need not be symmetric about their major diagonal.
	The row-column format of a matrix corresponds to query-subject letter pairs.
	<BR>
	<I>See also:</I>
	<CODE><A HREF="#altscore">altscore</A></CODE>,
	<CODE><A HREF="#m">M</A></CODE> and <CODE><A HREF="#n">N</A></CODE>.
	</TD>
	</TR>
	<TR>
		<TD>
		&nbsp;&nbsp;<CODE>mformat=</CODE>&lt;<I>m</I>&gt;[,<I>outfile</I>]</TD>
		<TD><A NAME="mformat"></A>
<!-- PARM="mformat" -->
		used to select an output format by numerical identifier, <I>m</I>, and optionally
		the name of the file where the output should be written, <I>outfile</I>.
		Multiple formats (multiple <CODE>mformat</CODE> options) may be requested for simultaneous output during a single search,
		as long as a different <I>outfile</I> is indicated for each format.
		If no <I>outfile</I> is specified, either standard output
		(<CODE>stdout</CODE>)
		or the setting of the <CODE>O</CODE> option (if set) is used.
		At most one <CODE>mformat</CODE> specification on a given command line
		may lack an <I>outfile</I>.
		If <I>outfile</I> contains any white space (<I>e.g.,</I> blanks or tabs),
		the entire token should be enclosed in quotes, to prevent command line interpreters
		from breaking it into separate arguments.
		<BR>
		The various output formats available are displayed if
		<CODE>mformat=list</CODE> is specified
		on an otherwise syntactically correct command line.
		(Example:&nbsp;&nbsp;<CODE>blastp&nbsp;foo&nbsp;foo&nbsp;mformat=list</CODE>).
		Setting <CODE>mformat=0</CODE> clears any <CODE>mformat</CODE>
		specification(s) appearing to the left on the command line.
		<BR>
		Depending on the output format, some command line options cause additional elements to appear;
		these options include:  <CODE><A HREF="#topcombon">topcomboN</A></CODE>, <CODE><A HREF="#topcomboe">topcomboE</A></CODE> and <CODE><A HREF="#links">links</A></CODE>.
		<P>
		The available choices for <I>m</I> and their associated formats are:
		</P>
		<P>
    <B>list</B> output this list and halt;
    <BR>
    <B>0</B> reset to default output only;
    <BR>
    <B>1</B> pairwise (default);
    <BR>
    <B>2</B> tabular (see <A HREF="http://blast.wustl.edu/blast/tabular.html">description</A>);
    <BR>
    <B>3</B> tabular with comments (see <A HREF="http://blast.wustl.edu/blast/tabular.html">description</A>);
    <BR>
    <B>4</B> PostScript&#8482; graphics<FONT COLOR="red">*</FONT> (see <A HREF="http://blast.wustl.edu/blast/ps.html">description</A>);
    <BR>
    <B>5</B> neighborhood word listings;<FONT COLOR="red">*</FONT>
    <BR>
    <B>7</B> XML conforming to <A HREF="http://blast.wustl.edu/dtd" TARGET="_blank">NCBI_BlastOutput.dtd</A> (see <A HREF="http://blast.wustl.edu/blast/xml-example.xml" TARGET="_blank">example</A>, best viewed in <A HREF="http://www.mozilla.org" TARGET="_blank">Firefox</A>).
    </P>
    <P>
		<FONT COLOR="red">*</FONT>Formats that are subject to change or removal without notice.
		</P>
		<I>See also:</I>
		<CODE><A HREF="#msgstyle">msgstyle</A></CODE>,
		<CODE><A HREF="#o">O</A></CODE> and
		<CODE><A HREF="#xmlcompact">xmlcompact</A></CODE>.
		</TD>
	</TR>
	<TR>
		<TD>
		&nbsp;&nbsp;<CODE>mmio</CODE></TD>
		<TD><A NAME="mmio"></A>
<!-- PARM="mmio" -->
		turns off the use of memory-mapped I/O when reading database files.
		Use of this option will usually slow the search, particularly when multiple processors are being used, but it serves both to demonstrate the effectiveness of this form of I/O and to validate the associated I/O routines.  Note that no special daemon or support programs (such as the old <I>memfile</I> program) are required to take full advantage of memory-mapped I/O.
		When running 32-bit versions of the BLAST software,
		the <CODE>mmio</CODE> option might free up important virtual address
		space for use as working storage or heap memory.
		<BR>
		<I>For the vast majority of users, this option should never be used.</I>
		</TD>
	</TR>
	<TR>
		<TD>
		&nbsp;&nbsp;<CODE>msgstyle=</CODE>&lt;<I>n</I>&gt;</TD>
		<TD><A NAME="msgstyle"></A>
<!-- PARM="msgstyle" -->
		used to select by numerical identifier, <I>n</I>, the style of informatory messages to produce (<I>i.e.</I>, <CODE>NOTE</CODE>s, <CODE>WARNING</CODE>s, etc.)
		<BR>
		The available choices for <I>n</I> and their associated styles are:
		<PRE>
    0 => line-wrapped (default)
    1 => single-line with the query sequence identifier embedded (if available)
		</PRE>
		</TD>
	</TR>
	<TR>
		<TD>
		&nbsp;&nbsp;<CODE>N=</CODE>&lt;<I>n</I>&gt;</TD>
	<TD><A NAME="n"></A>
<!-- PARM="N" -->
	set the negative penalty score for mismatching nucleotides
	in the <TT>BLASTN</TT> search mode to <I>n</I>, with default value -4.
		<BR>
		<I>See also:</I>
		<CODE><A HREF="#m">M</A></CODE>,
		<CODE><A HREF="#matrix">matrix</A></CODE>, and
		<CODE><A HREF="#altscore">altscore</A></CODE>.
	</TD>
	</TR>
	<TR>
		<TD>
		&nbsp;&nbsp;<CODE>nogaps</CODE></TD>
		<TD><A NAME="nogaps"></A>
<!-- PARM="nogaps" -->
		do not create gapped alignments, in essence reverting to WU&nbsp;BLAST&nbsp;1.4 behavior
		<BR>
		<I>See also:</I> <CODE><A HREF="#gaps">gaps</A></CODE> and <CODE><A HREF="#gapall">gapall</A></CODE>.
		</TD>
	</TR>
	<TR>
		<TD>
		&nbsp;&nbsp;<CODE>nonnegok</CODE></TD>
		<TD><A NAME="nonnegok"></A>
<!-- PARM="nonnegok" -->
		Do not abort processing with a <CODE>FATAL</CODE> error when the expected score
		is non-negative.
		Formally, for Karlin-Dembo-Altschul statistics to apply to the
		evaluation of the alignment scores found during a search,
		the expected score for a sequence having the same residue composition
		as the query must be negative, but this condition does not always
		hold with unusual scoring matrices or query sequences.
		Use the <CODE>novalidctxok</CODE> option to cause the search to proceed
		even under these unusual conditions.
		<BR>
		<I>See also:</I> <CODE><A HREF="#novalidctxok">novalidctxok</A></CODE> and <CODE><A HREF="#shortqueryok">shortqueryok</A></CODE>.
		</TD>
	</TR>
	<TR>
		<TD>
		&nbsp;&nbsp;<CODE>nosegs</CODE></TD>
		<TD><A NAME="nosegs"></A>
<!-- PARM="nosegs" -->
		do not segment the query sequence on hyphens (-).
		By default, hyphens in the query sequence create insurmountable
		barriers for sequence alignment.
		As an example of where this feature is useful,
		multiple contigs may be concatenated together into one sequence
		with a hyphen separating each contig;
		no alignment will then extend beyond a contig boundary.
		<BR>
		CAUTION: do not confuse this option with the similarly appearing
		<CODE><A HREF="#noseqs">noseqs</A></CODE> option.
		</TD>
	</TR>
	<TR>
		<TD>
		&nbsp;&nbsp;<CODE>noseqs</CODE></TD>
		<TD><A NAME="noseqs"></A>
<!-- PARM="noseqs" -->
		produce abbreviated output by omitting the sequence alignments.
		The result is often correctly interpretable by parsers of normal output.
		<BR>
		CAUTION: do not confuse this option with the similarly appearing
		<CODE><A HREF="#nosegs">nosegs</A></CODE> option.
		</TD>
	</TR>
	<TR>
		<TD>
		&nbsp;&nbsp;<CODE>notes</CODE></TD>
		<TD><A NAME="notes"></A>
<!-- PARM="notes" -->
		suppress all <CODE>NOTE</CODE> messages.  Important recommendations from the software may be missed if this option is used.
		If any <CODE>NOTE</CODE>s arise with this option, the number <CODE>SUPPRESSED</CODE> will be reported at the end of the search.
		<BR>
		<I>See also:</I>
		<CODE><A HREF="#errors">errors</A></CODE>
		and
		<CODE><A HREF="#warnings">warnings</A></CODE>.
		</TD>
	</TR>
	<TR>
		<TD>
		&nbsp;&nbsp;<CODE>novalidctxok</CODE></TD>
		<TD><A NAME="novalidctxok"></A>
<!-- PARM="novalidctxok" -->
		do not treat it as a <CODE>FATAL</CODE> error when none of the &ldquo;contexts&rdquo;
		(<I>e.g.,</I> strands or reading frames) of the query are valid.
		A valid context is one in which the threshold score for saving
		alignments can be achieved under ideal circumstances (typically
		if an alignment of 100% identity were to be found).
		<BR>
		<I>See also:</I> <CODE><A HREF="#nonnegok">nonnegok</A></CODE> and <CODE><A HREF="#shortqueryok">shortqueryok</A></CODE>.
		</TD>
	</TR>
	<TR>
		<TD>
		&nbsp;&nbsp;<CODE>nwlen=</CODE>&lt;<I>len</I>&gt;</TD>
		<TD><A NAME="nwlen"></A>
<!-- PARM="nwlen" -->
		generate neighborhood words (or seed words) starting from
		the beginning of the query sequence (or from the location specified
		with the <CODE>nwstart</CODE> parameter) and continuing
		for the distance <I>len</I> or to the end of the sequence,
		whichever comes first.
		While this parameter can be used to restrict the region in which
		word hits occur for seeding ungapped alignments (and indirectly gapped alignments),
		it does not restrict alignments from extending beyond this region.
		<BR>
		<I>See also:</I> <CODE><A HREF="#nwstart">nwstart</A></CODE>.
		</TD>
	</TR>
	<TR>
		<TD>
		&nbsp;&nbsp;<CODE>nwstart=</CODE>&lt;<I>start</I>&gt;</TD>
		<TD><A NAME="nwstart"></A>
<!-- PARM="nwstart" -->
		generate neighborhood words (or seed words) starting from
		coordinate position <I>start</I> in the query sequence and continuing
		to the end of the sequence (or for the distance specified with the <CODE>nwlen</CODE> parameter).
		While this parameter can be used to restrict the region in which
		word hits occur for seeding ungapped alignments (and indirectly gapped alignments),
		it does not restrict alignments from extending beyond this region.
		<BR>
		<I>See also:</I> <CODE><A HREF="#nwlen">nwlen</A></CODE>.
		</TD>
	</TR>
	<TR>
		<TD>
		&nbsp;&nbsp;<CODE>O=</CODE>&lt;<I>outfile</I>&gt;</TD>
		<TD><A NAME="o"></A>
<!-- PARM="O" -->
		output results to the file named <I>outfile</I> instead of standard output (<CODE>stdout</CODE>).
		</TD>
	</TR>
	<TR>
		<TD>
		&nbsp;&nbsp;<CODE>olfraction=</CODE>&lt;<I>f</I>&gt;</TD>
		<TD><A NAME="olfraction"></A>
<!-- PARM="olfraction" -->
		set the maximum fractional length of overlap, <I>f</I>, of two <I>un</I>gapped alignments
		for them to be considered independent and mutually &ldquo;consistent&rdquo; and their joint (Sum or Poisson) probability to be computed.
		The default <I>f</I> is 0.1 (maximum 10% of the length from either end of either HSP).
		For any given pair of HSPs, the more restrictive of <CODE>olfraction</CODE>
		and <CODE>olmax</CODE> is used.
		To eliminate <CODE>olfraction</CODE> from consideration,
		set its value to 1,
		to indicate the acceptability of even a complete, 100% overlap.
		<BR>
		<I>See also:</I>
		<CODE><A HREF="#golfraction">golfraction</A></CODE>,
		<CODE><A HREF="#golmax">golmax</A></CODE>,
		and
		<CODE><A HREF="#olmax">olmax</A></CODE>.
		</TD>
	</TR>
	<TR>
		<TD>
		&nbsp;&nbsp;<CODE>olmax=</CODE>&lt;<I>len</I>&gt;</TD>
		<TD><A NAME="olmax"></A>
<!-- PARM="olmax" -->
		set the maximum permitted length of overlap (in residues), <I>len</I>, of two <I>un</I>gapped alignments
		for their joint (Sum or Poisson) probability to be computed.
		The default is unlimited length, with the maximum extent of overlap being governed only
		by the <CODE>olfraction</CODE> parameter.
		<BR>
		<I>See also:</I>
		<CODE><A HREF="#golfraction">golfraction</A></CODE>,
		<CODE><A HREF="#golmax">golmax</A></CODE>,
		and
		<CODE><A HREF="#olfraction">olfraction</A></CODE>.
		</TD>
	</TR>
	<TR>
		<TD>
		&nbsp;&nbsp;<CODE>pingpong</CODE></TD>
		<TD><A NAME="pingpong"></A>
<!-- PARM="pingpong" -->
		Perform additional work to help ensure the alignments produced are locally optimal.
		This option typically adds 3-10% to the execution time, without affecting
		the results.
		Only rarely is an alignment and its associated score improved, for the additional time consumed by using this option.
		</TD>
	</TR>
	<TR>
		<TD>
		&nbsp;&nbsp;<CODE>poissonp</CODE></TD>
		<TD><A NAME="poissonp"></A>
<!-- PARM="poissonp" -->
		use Poisson statistics
		(<A HREF="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=Retrieve&amp;db=pubmed&amp;dopt=Abstract&amp;list_uids=2315319">Karlin and Altschul, 1990</A>)
		to compute joint <I>P</I>-values of consistent sets of alignments;
		Poisson statistics are an option in all search modes.</TD>
	</TR>
	<TR>
		<TD>
		&nbsp;&nbsp;<CODE>postsw</CODE></TD>
		<TD><A NAME="postsw"></A>
<!-- PARM="postsw" -->
		perform full Smith-Waterman alignment of sequences and re-rank the database matches accordingly prior to output (currently supported in <TT>BLASTP</TT> only)</TD>
	</TR>
	<TR>
		<TD>
		&nbsp;&nbsp;<CODE>progress=</CODE>&lt;<I>s</I>&gt;</TD>
		<TD><A NAME="progress"></A>
<!-- PARM="progress" -->
		provide an indication that the search is alive by outputting an asterisk (&ldquo;*&rdquo;) every <I>s</I> seconds during a search,
		if some other indication of activity has not been provided in the mean time.
		Such &ldquo;keepalive&rdquo; indicators may be useful when the software
		is invoked over a network connection.
		The default behavior
		(obtained with <CODE>progress=0</CODE>)
		is only to report the actual progress made through the database,
		using periods (&ldquo;.&rdquo;) and reports of percentages.
		</TD>
	</TR>
	<TR>
		<TD>
		&nbsp;&nbsp;<CODE>prune</CODE></TD>
		<TD><A NAME="prune"></A>
<!-- PARM="prune" -->
		do <I>not</I> prune HSP lists, but instead report all HSPs, even
		those that were not involved
		in satisfying the statistical significance threshold necessary
		for reporting the database sequence.
		NOTE: When the default Sum statistics are used,
		the normal pruning activity is robust;
		when Poisson statistics are used,
		some HSPs may get through the pruning process and be reported
		that were not involved
		in satisfying the statistical significance threshold.
		<BR>
		<I>See also:</I> <CODE><A HREF="#span">span</A></CODE>,
		<CODE><A HREF="#span1">span1</A></CODE> and
		<CODE><A HREF="#span2">span2</A></CODE>.
		</TD>
	</TR>
	<TR>
		<TD>
		&nbsp;&nbsp;<CODE>putenv=</CODE>"<I><FONT SIZE=-1>NAME=VALUE</FONT></I>"</TD>
		<TD><A NAME="putenv"></A>
<!-- PARM="putenv" -->
		in the local environment to the BLAST search program, set the environment variable named <I><FONT SIZE=-1>NAME</FONT></I> to the value <I><FONT SIZE=-1>VALUE</FONT></I>.
		<BR>
		<I>See also:</I> <CODE><A HREF="#endgetenv">endgetenv</A></CODE>,
		<CODE><A HREF="#endputenv">endputenv</A></CODE> and <CODE><A HREF="#getenv">getenv</A></CODE>.
		</TD>
	</TR>
	<TR>
		<TD>
		&nbsp;&nbsp;<CODE>pvalues</CODE></TD>
		<TD><A NAME="pvalues"></A>
<!-- PARM="pvalues" -->
		report <I>P</I>-values (the default) in the initial one-line descriptions section of output.
		<BR>
		<I>See also:</I> <CODE><A HREF="#evalues">evalues</A></CODE>.
		</TD>
	</TR>
	<TR>
		<TD>
		&nbsp;&nbsp;<CODE>Q=</CODE>&lt;<I>q</I>&gt;</TD>
		<TD><A NAME="q"></A>
<!-- PARM="Q" -->
		set the penalty for a gap of length one to <I>q</I> (default Q=9 for proteins; Q=10 for <TT>BLASTN</TT>).
		<BR>
		<I>See also:</I> <CODE><A HREF="#r">R</A></CODE>.</TD>
	</TR>
	<TR>
		<TD>
		&nbsp;&nbsp;<CODE>qframe=</CODE>&lt;<I>f</I>&gt;</TD>
		<TD><A NAME="qframe"></A>
<!-- PARM="qframe" -->
		search with the query sequence translated in the single reading frame <I>f</I>.
		This parameter is useful for speeding up a search and improving both the
		biological and statistical significance of the findings,
		when the reading frame of a translation product in the query
		is known in advance,
		such as when the query sequence entails a complete ORF.
		Reading frames on the top (plus) strand of the query are numbered 1, 2, 3;
		reading frames on the bottom (minus) strand are numbered -1, -2, -3.
		<BR>
		<I>See also:</I>
		<CODE><A HREF="#top">top</A></CODE>,
		<CODE><A HREF="#bottom">bottom</A></CODE>,
		<CODE><A HREF="#dbtop">dbtop</A></CODE> and
		<CODE><A HREF="#dbbottom">dbbottom</A></CODE>.
		</TD>
	</TR>
	<TR>
		<TD>
		&nbsp;&nbsp;<CODE>Qoffset=</CODE>&lt;<I>i</I>&gt;</TD>
		<TD><A NAME="qoffset"></A>
<!-- PARM="qoffset" -->
		adjust all query sequence coordinates in the output by the fixed
		quantity <I>i</I> (default 0).
		</TD>
	</TR>
	<TR>
		<TD>
		&nbsp;&nbsp;<CODE>qrecmax=</CODE>&lt;<I>n</I>&gt;</TD>
		<TD><A NAME="qrecmax"></A>
<!-- PARM="qrecmax" -->
		in a multi-sequence query file, end database searches with the query sequence numbered <I>n</I>.</TD>
	</TR>
	<TR>
		<TD>
		&nbsp;&nbsp;<CODE>qrecmin=</CODE>&lt;<I>m</I>&gt;</TD>
		<TD><A NAME="qrecmin"></A>
<!-- PARM="qrecmin" -->
		in a multi-sequence query file, start database searches using the query sequence numbered <I>m</I>.  Record are numbered starting with 1.</TD>
	</TR>
	<TR>
		<TD>
		&nbsp;&nbsp;<CODE>qres</CODE></TD>
		<TD><A NAME="qres"></A>
<!-- PARM="qres" -->
		treat as a <CODE>FATAL</CODE> error when the query sequence contains any invalid residue codes.
		By default, <CODE>WARNING</CODE>s are issued for invalid residue codes,
		which are then skipped.
		</TD>
	</TR>
	<TR>
		<TD>
		&nbsp;&nbsp;<CODE>qtype</CODE></TD>
		<TD><A NAME="qtype"></A>
<!-- PARM="qtype" -->
		treat as a <CODE>FATAL</CODE> error if the query sequence appears from its letter composition to be of the wrong type (peptide or nucleotide).
		</TD>
	</TR>
	<TR>
		<TD>
		&nbsp;&nbsp;<CODE>R=</CODE>&lt;<I>r</I>&gt;</TD>
		<TD><A NAME="r"></A>
<!-- PARM="R" -->
		set the per-residue penalty for extending a gap to <I>r</I> (default R=2 for proteins; R=10 for <TT>BLASTN</TT>)
		<BR>
		<I>See also:</I> <CODE><A HREF="#q">Q</A></CODE>.
		</TD>
	</TR>
	<TR>
		<TD>
		&nbsp;&nbsp;<CODE>restest</CODE></TD>
		<TD><A NAME="restest"></A>
<!-- PARM="restest" -->
		causes the Bonferroni corrections used in computing
		statistical significance to depend
		upon the length in residues of each database sequence
		relative to the total number of <I>residues</I> in the database.
		<CODE>restest</CODE>
		is the default database-size correction method
		in the <TT>BLASTN</TT>, <TT>TBLASTN</TT>, and <TT>TBLASTX</TT> search modes.
		<BR>
		<I>See also:</I> <CODE><A HREF="#seqtest">seqtest</A></CODE>.
		</TD>
	</TR>
	<TR>
		<TD>
		&nbsp;&nbsp;<CODE>S=</CODE>&lt;<I>s</I>&gt;</TD>
		<TD><A NAME="s"></A>
<!-- PARM="S" -->
	set the score-equivalence threshold for reporting database hits to <I>s</I>.
	Hits for a database sequence will only be reported
	if the statistical significance ascribed to one of its similar
	regions (or groups of similar regions)
	is at least as high as that of a <I>single</I> alignment with score
	<CODE>S</CODE>.
	By default, <CODE>S</CODE> is not actually used;
	all findings are compared against the <CODE>E</CODE> threshold.
	<CODE>E</CODE> and <CODE>S</CODE> are interchangeable, however,
	through standard Karlin-Dembo-Altschul statistics,
	with any setting of <CODE>S</CODE> implying
	an expectation threshold of its own
	(<I>E=KN</I>e<sup><I>-&lambda;S</I></sup>).
	If both <CODE>E</CODE> and <CODE>S</CODE> are specified
	on the command line, the one corresponding to the more restrictive (lower)
	<I>E</I> is used.
		<BR>
		<I>See also:</I>
		<CODE><A HREF="#e">E</A></CODE>,
		<CODE><A HREF="#gaps2">gapS2</A></CODE>,
		and
		<CODE><A HREF="#s2">S2</A></CODE>.
	</TD>
	</TR>
	<TR>
		<TD>
		&nbsp;&nbsp;<CODE>S2=</CODE>&lt;<I>s</I>&gt;</TD>
	<TD><A NAME="s2"></A>
<!-- PARM="S2" -->
	set the score threshold for saving ungapped HSPs to <I>s</I>.
	In the initial, ungapped alignment phase of a search,
	individual HSPs will only be saved for further use
	if their score is &ge;<CODE>S2</CODE>.
	The default score threshold is computed from <CODE>E2</CODE>
	and will depend on the scoring system.
	If both <CODE>E2</CODE> and <CODE>S2</CODE> are specified on the command line,
	the one corresponding to the more restrictive (higher) score threshold
	will be used.
		<BR>
		<CODE><A HREF="#gaps2">gapS2</A></CODE>,
		<CODE><A HREF="#e2">E2</A></CODE> and
		<CODE><A HREF="#gape2">gapE2</A></CODE>.
	</TD>
	</TR>
	<TR>
		<TD>
		&nbsp;&nbsp;<CODE>seqtest</CODE></TD>
		<TD><A NAME="seqtest"></A>
<!-- PARM="seqtest" -->
		causes the Bonferroni corrections used in computing
		statistical significance to depend upon
		the number of <I>sequences</I> in the database.
		<CODE>seqtest</CODE> is the default database-size correction method
		in the <TT>BLASTP</TT> and <TT>BLASTX</TT> search modes.
		<BR>
		<TT>NOTE:</TT> For backward compatibility with legacy <TT>BLAST</TT> software,
		in all search modes, including <TT>BLASTP</TT> and <TT>BLASTX</TT>,
		if the <CODE>Z</CODE> option is specified,
		<CODE>Z</CODE> is expected to be expressed in units of residues,
		unless <CODE>seqtest</CODE> is also specified.
		<BR>
		<I>See also:</I> <CODE><A HREF="#restest">restest</A></CODE> and
		<CODE><A HREF="#z">Z</A></CODE>.
		</TD>
	</TR>
	<TR>
		<TD>
		&nbsp;&nbsp;<CODE>shortqueryok</CODE></TD>
		<TD><A NAME="shortqueryok"></A>
<!-- PARM="shortqueryok" -->
		do not treat it as a <CODE>FATAL</CODE> error when the query sequence is
		shorter than the <TT>BLAST</TT> algorithm word length.
		<BR>
		<I>See also:</I> <CODE><A HREF="#novalidctxok">novalidctxok</A></CODE> and
		<CODE><A HREF="#nonnegok">nonnegok</A></CODE>.</TD>
	</TR>
	<TR>
		<TD>
		&nbsp;&nbsp;<CODE>Soffset=</CODE>&lt;<I>i</I>&gt;</TD>
		<TD><A NAME="soffset"></A>
<!-- PARM="soffset" -->
		adjust all subject sequence coordinates in the output by the fixed
		quantity <I>i</I> (default 0).
		</TD>
	</TR>
	<TR>
		<TD>
		&nbsp;&nbsp;<CODE>sort_by_count</CODE></TD>
		<TD><A NAME="sort_by_count"></A>
<!-- PARM="sort_by_count" -->
		sort database sequences from highest to lowest by the number of HSPs identified.
		Multiple <CODE>sort_by*</CODE> options may be specified and take precedence in the order specified.
		</TD>
	</TR>
	<TR>
		<TD>
		&nbsp;&nbsp;<CODE>sort_by_highscore</CODE></TD>
		<TD><A NAME="sort_by_highscore"></A>
<!-- PARM="sort_by_highscore" -->
		sort database sequences from highest to lowest by the highest HSP score found.
		Multiple <CODE>sort_by*</CODE> options may be specified and take precedence in the order specified.
		</TD>
	</TR>
	<TR>
		<TD>
		&nbsp;&nbsp;<CODE>sort_by_pvalue</CODE></TD>
		<TD><A NAME="sort_by_pvalue"></A>
<!-- PARM="sort_by_pvalue" -->
		sort database sequences from lowest to highest by their best <I>P</I>-value.
		Multiple <CODE>sort_by*</CODE> options may be specified
		and take precedence in the order specified.
		<CODE>sort_by_pvalue</CODE> is the default primary sort key.
		</TD>
	</TR>
	<TR>
		<TD>
		&nbsp;&nbsp;<CODE>sort_by_subjectlength</CODE></TD>
		<TD><A NAME="sort_by_subjectlength"></A>
<!-- PARM="sort_by_subjectlength" -->
		sort database sequences from longest to shortest.
		Multiple <CODE>sort_by*</CODE> options may be specified and take precedence in the order specified.
		</TD>
	</TR>
	<TR>
		<TD>
		&nbsp;&nbsp;<CODE>sort_by_totalscore</CODE></TD>
		<TD><A NAME="sort_by_totalscore"></A>
<!-- PARM="sort_by_totalscore" -->
		sort database sequences from highest to lowest by the sum total score of all HSPs found.
		Multiple <CODE>sort_by*</CODE> options may be specified and take precedence in the order specified.
		</TD>
	</TR>
	<TR>
		<TD>
		&nbsp;&nbsp;<CODE>span</CODE></TD>
		<TD><A NAME="span"></A>
<!-- PARM="span" -->
		retain HSPs (ungapped or gapped) regardless of whether they
		span or are spanned by any other HSP.
		When this option is specified, memory requirements may increase
		dramatically to accommodate an increased number of HSPs that must
		be tracked, particularly when the sequences being compared
		contain short periodicity repeats and low complexity regions.
		<BR>
		<I>See also:</I> <CODE><A HREF="#span1">span1</A></CODE> and <CODE><A HREF="#span2">span2</A></CODE>.
		</TD>
	</TR>
	<TR>
		<TD>
		&nbsp;&nbsp;<CODE>span1</CODE></TD>
		<TD><A NAME="span1"></A>
<!-- PARM="span1" -->
		discard an HSP (ungapped or gapped) when it spans or is spanned by
		another HSP along <I>either</I> the query or the subject sequence (or both).
		When a pair of such HSPs is found, the one with the lowest score
		is discarded;
		if their scores are equal, the longer, less information-dense HSP is discarded.
		<BR>
		<I>See also:</I> <CODE><A HREF="#span">span</A></CODE> and <CODE><A HREF="#span2">span2</A></CODE>.
		</TD>
	</TR>
	<TR>
		<TD>
		&nbsp;&nbsp;<CODE>span2</CODE></TD>
		<TD><A NAME="span2"></A>
<!-- PARM="span2" -->
		discard an HSP (ungapped or gapped) when it spans or is spanned by
		another HSP along <I>both</I> the query and subject sequences.
		When a pair of such HSPs is found, the one with the lowest score
		is discarded; if their scores are equal, the longer, less information-dense HSP is discarded.
		<BR>
		<CODE>span2</CODE> is the default behavior.
		<BR>
		<I>See also:</I> <CODE><A HREF="#span">span</A></CODE> and <CODE><A HREF="#span1">span1</A></CODE>.
		</TD>
	</TR>
	<TR>
		<TD>
		&nbsp;&nbsp;<CODE>spoutmax=</CODE>&lt;<I>spoutmax</I>&gt;</TD>
		<TD><A NAME="spoutmax"></A>
<!-- PARM="spoutmax" -->
		establishes <I>spoutmax</I>
		as the maximum number of segment pairs to report
		in program output per subject sequence or pairwise comparison,
		independent of the number of HSPs or GSPs actually found and evaluated.
		If more than <CODE>spoutmax</CODE> segment pairs are found,
		the segment pairs are sorted by the criteria in effect
		for the search and only the first <CODE>spoutmax</CODE>
		segment pairs will be reported.
		The setting of <CODE>spoutmax</CODE> will have no effect
		if either <CODE>hspmax</CODE> or <CODE>gspmax</CODE>
		is more restrictive.
		<BR>
		The default value for <CODE>spoutmax</CODE> is 0,
		which signifies no limit.
		<BR>
		<I>See also:</I>
		<CODE><A HREF="#hspmax">hspmax</A></CODE> and
		<CODE><A HREF="#gspmax">gspmax</A></CODE>.
		</TD>
	</TR>
	<TR>
		<TD>
		&nbsp;&nbsp;<CODE>stats</CODE></TD>
		<TD><A NAME="stats"></A>
<!-- PARM="stats" -->
		gather a variety of statistics about the search
		(<I>e.g.,</I> the number of word hits in each reading frame,
		the highest score observed, etc.)
		and report them in the output.
		Use of this option marginally impacts search speed.
		</TD>
	</TR>
	<TR>
		<TD>
		&nbsp;&nbsp;<CODE>sump</CODE></TD>
		<TD><A NAME="sump"></A>
<!-- PARM="sump" -->
		use Sum statistics
		(<A HREF="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=Retrieve&amp;db=pubmed&amp;dopt=Abstract&amp;list_uids=8390686">Karlin and Altschul, 1993</A>)
		to compute joint <I>P</I>-values of consistent sets of alignments;
		the use of Sum statistics is the default behavior
		in all search modes.</TD>
	</TR>
	<TR>
		<TD>
		&nbsp;&nbsp;<CODE>T=</CODE>&lt;<I>t</I>&gt;</TD>
	<TD><A NAME="t"></A>
<!-- PARM="T" -->
	set the neighborhood word score threshold for the ungapped <TT>BLAST</TT> algorithm to <I>t</I>.
	For a given word of length <CODE>W</CODE> in the query sequence,
	its neighborhood words are defined as the set of words
	that have scores &ge;&nbsp;<CODE>T</CODE> when aligned with it.
	Neighborhood words become the seed words used to find ungapped alignments
	by the <TT>BLAST</TT> algorithm.
	Lower values for <CODE>T</CODE> tend to yield a larger neighborhood,
	more potential seed words, and improved sensitivity for lower scoring alignments,
	but at the expense of increased memory use and run time.
	Higher values for <CODE>T</CODE> will yield a smaller (possibly empty) neighborhood word list and faster execution, at the expense of reduced sensitivity.
	The default <CODE>T</CODE> varies with the scoring matrix, word length, and
	between search modes.
	For improved sensitivity and to obtain behavior that better satisfies user expectations,
	identical words are included with neighborhood words in the list
	of potential seeds,
	if their score is positive but happens to be less than <CODE>T</CODE>.
	<BR>
	No neighborhood words (only exactly matching words) are used by default
	in the <TT>BLASTN</TT> search mode;
	however, neighborhood words can be used even by <TT>BLASTN</TT> if a value
	for <CODE>T</CODE> is specified on the <TT>BLASTN</TT> command line.
	<TT>CAUTION</TT>:  for the long word lengths typically employed
	with <TT>BLASTN</TT>, the memory required
	for neighborhood words can easily be prohibitive and may only be
	practical for shorter sequences.
	</TD>
	</TR>
	<TR>
		<TD>
		&nbsp;&nbsp;<CODE>top</CODE></TD>
		<TD><A NAME="top"></A>
<!-- PARM="top" -->
		used to restrict the search of a nucleotide sequence
		to the top (+) strand.
		In the <TT>TBLASTX</TT> search mode, where both query and subject
		are nucleotide sequences, the <CODE>top</CODE> option only affects
		the query sequence.
		<BR>
		<I>See also:</I> <CODE><A HREF="#bottom">bottom</A></CODE>,
		<CODE><A HREF="#dbtop">dbtop</A></CODE>,
		<CODE><A HREF="#dbbottom">dbbottom</A></CODE> and
		<CODE><A HREF="#qframe">qframe</A></CODE>.
		</TD>
	</TR>
	<TR>
		<TD>
		&nbsp;&nbsp;<CODE>topcomboE=</CODE>&lt;<I>E<sub>ratio</sub></I>&gt;</TD>
		<TD><A NAME="topcomboe"></A>
<!-- PARM="topcomboE" -->
		<I>E<sub>ratio</sub></I> is the maximum ratio of <I>E<sub>current</sub>/E<sub>best</sub></I> for which
		the <I>current</I> &ldquo;topcombo&rdquo; group of consistent (colinear) local alignments will be reported
		for a given database sequence.
		The "best" group is reported in the output as "Group = 1"
		and tends to be the most statistically significant.
		The default behavior is to impose no limit on this ratio, in which case all topcombo groups satisfying E are reported (up to a maximum of <CODE>topcomboN</CODE> groups, if specified).
		<BR>
		<I>See also:</I> <CODE><A HREF="#links">links</A></CODE> and
		<CODE><A HREF="#topcombon">topcomboN</A></CODE>.
		</TD>
	</TR>
	<TR>
		<TD>
		&nbsp;&nbsp;<CODE>topcomboN=</CODE>&lt;<I>n</I>&gt;</TD>
		<TD><A NAME="topcombon"></A>
<!-- PARM="topcomboN" -->
		report at most <I>n</I> &ldquo;topcombo&rdquo; groups of consistent (colinear) local alignments (HSPs).
		Each local alignment is allowed to be a member of only one group.
		Use of this option causes the addition of a "Group = #" indicator
		in the output for each HSP.
		Groups of HSPs tend to be assembled in decreasing order of statistical
		significance.
		Members of the most significant group thus tend to be reported
		with "Group = 1".
		<BR>
		<I>See also:</I> <CODE><A HREF="#links">links</A></CODE> and
		<CODE><A HREF="#topcomboe">topcomboE</A></CODE>.
		</TD>
	</TR>
	<TR>
		<TD>
		&nbsp;&nbsp;<CODE>ucdb</CODE></TD>
	<TD><A NAME="ucdb"></A>
<!-- PARM="ucdb" -->
	force nucleotide sequence databases to be searched in their uncompressed form,
	with any-and-all ambiguity codes in place.
	This option is only effective in the <TT>BLASTN</TT> search mode for word lengths &ge; 7.
	<I>Users should generally avoid specifying this option themselves,
	letting the software decide when to employ this search strategy.</I>
	This option can increase sensitivity when ambiguity codes
	are present in database sequences,
	at the expense of memory and possibly speed.
	Searching the uncompressed database is the only available behavior
	for word lengths &lt; 7.
	This option offers improved sensitivity only when searching databases in XDF format that contain ambiguity codes.
	The option is accepted by the software but offers no improvement in sensitivity for databases in the earlier BLAST&nbsp;1.4 database format.
		<BR>
		<I>See also:</I> <CODE><A HREF="#cdb">cdb</A></CODE>.
	</TD>
	</TR>
	<TR>
		<TD>
		&nbsp;&nbsp;<CODE>V=</CODE>&lt;<I>v</I>&gt;</TD>
	<TD><A NAME="v"></A>
<!-- PARM="V" -->
	set the maximum number of one-line descriptions of significant
	database sequences to report in the first section of program output to <I>v</I>.
	The default limit is 500.
		<BR>
		<I>See also:</I> <CODE><A HREF="#b">B</A></CODE>.
	</TD>
	</TR>
	<TR>
		<TD>
		&nbsp;&nbsp;<CODE>W=</CODE>&lt;<I>w</I>&gt;</TD>
	<TD><A NAME="w"></A>
<!-- PARM="W" -->
	set the seed word length for the ungapped <TT>BLAST</TT> algorithm to <I>w</I>.
	The default word length for protein-level searches is 3 amino acids;
	for <TT>BLASTN</TT> searches, the default length is 11 nucleotides.
	Shorter word lengths may increase sensitivity,
	at the expense of increased run time.
	In all search modes,
	the acceptable range of word lengths is 1&nbsp;&le;&nbsp;<I>w</I>&nbsp;&le;&nbsp;1024.
	</TD>
	</TR>
	<TR>
		<TD>
		&nbsp;&nbsp;<CODE>warnings</CODE></TD>
		<TD><A NAME="warnings"></A>
<!-- PARM="warnings" -->
		suppress all <CODE>WARNING</CODE> messages.
		<BR>
		<I><TT>CAUTION</TT></I>: important advisories may be missed if this option is used;
		however, if any <CODE>WARNING</CODE> situations should arise,
		the number <CODE>SUPPRESSED</CODE> will be reported at the end of the search.
		<BR>
		<I>See also:</I>
		<CODE><A HREF="#errors">errors</A></CODE>
		and
		<CODE><A HREF="#notes">notes</A></CODE>.
		</TD>
	</TR>
	<TR>
		<TD>
		&nbsp;&nbsp;<CODE>wink=</CODE>&lt;<I>wink</I>&gt;</TD>
		<TD><A NAME="wink"></A>
<!-- PARM="wink" -->
		generate word hits at every <I>wink</I><sup>th</sup> residue position along the query,
where the default <CODE>wink=1</CODE> produces neighborhood words at every position.
For best sensitivity, <CODE>wink</CODE> should not be adjusted.
<CODE>Wink</CODE> settings greater than 1 are best used to find identical or nearly identical sequences more rapidly.
When used in conjunction with the hitdist option
to obtain the highest search speed, care should be taken that desirable alignments are not precluded
by these parameters.
	The <CODE>wink</CODE> parameter is only available in the licensed 2.0 software.
	<BR>
	<I>NOTE:</I>
	When using <TT>BLASTN</TT> to search compressed nucleotide sequence
	databases in their compressed form,
	an increase in speed (and concommitant decrease in sensitivity)
	will not be observed unless <CODE>wink</CODE> is set to a value
	greater than the compression ratio, which is usually 4.
	<BR>
	<B>CAUTION:  With versions of <TT>BLASTN</TT> prior to [15-Oct-2004],
	similarity of any length and even 100% identity can be missed
	when searching compressed nucleotide sequence databases in their
	compressed form,
	if <CODE>wink</CODE> is set to an <I>even</I> integer value.
	This is simply due to the likelihood of a phase mismatch between the
	compressed form of the query and the database sequence.
	Assigning odd values to <CODE>wink</CODE> can avoid such phase mismatches.
	The best solution, though, is to update to a newer version of <TT>BLASTN</TT>.
	Versions of WU&nbsp;BLAST dated [15-Oct-2004] and later automatically
	avoid the phase mismatch problem, so users need not be concerned.
	If you are using the <CODE>wink</CODE> option with <TT>BLASTN</TT>
	and are not running a more recent version, please update!
	</B>
	</TD>
	</TR>
	<TR>
		<TD>
		&nbsp;&nbsp;<CODE>wordmask=</CODE>&lt;<I>filter</I>&gt;</TD>
		<TD><A NAME="wordmask"></A>
<!-- PARM="wordmask" -->
		&ldquo;soft mask&rdquo; the query sequence using the indicated <I>filter</I>.
		A copy of the query sequence is passed through the filter program
		and any letters converted by it to ambiguity codes
		are skipped during neighborhood word or seed word generation.
		Unlike the <CODE>filter</CODE> option,
		the query sequence itself remains unaltered and available for alignment.
		Usage of the <CODE>wordmask</CODE> parameter is otherwise identical to that of <CODE>filter</CODE>,
		with the same set of filtering methods available for use.
		<BR>
		<I>See also:</I>
		<CODE><A HREF="#filter">filter</A></CODE>,
		<CODE><A HREF="#lcmask">lcmask</A></CODE>,
		<CODE><A HREF="#lcfilter">lcfilter</A></CODE> and
		<CODE><A HREF="#maskextra">maskextra</A></CODE>.
		</TD>
	</TR>
	<TR>
		<TD>
		&nbsp;&nbsp;<CODE>wstrict</CODE></TD>
		<TD><A NAME="wstrict"></A>
<!-- PARM="wstrict" -->
		when searching a nucleotide database sequence
		that contains one or more ambiguous residues,
		require that every ungapped alignment found during the initial, ungapped phase of a search
		actually contain an identical word hit (in the usual case of <TT>BLASTN</TT> usage)
		or neighborhood word hit (in the case of <TT>TBLASTN</TT> and <TT>TBLASTX</TT>).
		The <CODE>wstrict</CODE> option has no effect whatsoever on <TT>BLASTX</TT>
		and has no effect on <TT>BLASTP</TT> when gapped alignments (the default)
		are to be produced.
		When ungapped alignments are the desired end product from <TT>BLASTP</TT>
		(<I>i.e.,</I> the <CODE>-nogaps</CODE> option is specified),
		<CODE>wstrict</CODE> will prevent the software from exhaustively
		searching diagonals that are found to contain HSPs in an effort
		to find other HSPs that would not be seeded by neighborhood word hits.
		</TD>
	</TR>
	<TR>
		<TD>
		&nbsp;&nbsp;<CODE>X=</CODE>&lt;<I>x</I>&gt;</TD>
	<TD><A NAME="x"></A>
<!-- PARM="X" -->
	set the drop-off score for the ungapped <TT>BLAST</TT> algorithm to <I>x</I>.
	Ungapped extension of initial neighborhood word hits or seed word hits
	between the query and subject sequences
	continues until the cumulative alignment score deteriorates
	from the maximum value seen thusfar during the extension by a quantity <CODE>X</CODE> or more.
	The default value for <CODE>X</CODE> is the score associated with 10 bits
	of significance (2<sup>-10</sup> &lt; 10<sup>-3</sup> probability) for protein-level searches or 20 bits
	of significance (2<sup>-20</sup> &lt; 10<sup>-6</sup> probability)
	for nucleotide-level (<TT>BLASTN</TT>) searches.
	Higher values for <CODE>X</CODE> will increase sensitivity at the expense
	of run time, but with both typically diminishing rapidly in their rate of change.
		<BR>
		<I>See also:</I> <CODE><A HREF="#gapx">gapX</A></CODE>.
	</TD>
	</TR>
	<TR>
		<TD>
		&nbsp;&nbsp;<CODE>xmlcompact</CODE></TD>
	<TD><A NAME="xmlcompact"></A>
<!-- PARM="xmlcompact" -->
	omit newline and white space characters normally reported between
	entities in XML documents produced with <CODE>mformat=7</CODE>.
	Their purpose is merely to improve human readability of a document
	when viewed with an XML-ignorant program,
	but these characters often comprise
	a substantial fraction (30% is not uncommon) of the bytes in a document and they are completely
	extraneous for the purposes of automated parsing and viewing
	with XML-aware software.
		<BR>
		<I>See also:</I> <CODE><A HREF="#mformat">mformat</A></CODE>.
	</TD>
	</TR>
	<TR>
		<TD>
		&nbsp;&nbsp;<CODE>Y=</CODE>&lt;<I>y</I>&gt;</TD>
	<TD><A NAME="y"></A>
<!-- PARM="Y" -->
	set the effective length of the quer<I>Y</I> sequence (in units of residues)
	used in statistical significance calculations to <I>y</I>.
	</TD>
	</TR>
	<TR>
		<TD>
		&nbsp;&nbsp;<CODE>Z=</CODE>&lt;<I>z</I>&gt;</TD>
	<TD><A NAME="z"></A>
<!-- PARM="Z" -->
	set the effective size of the database (databa<I>Z</I>e) used in statistical significance calculations to <I>z</I>.
	Unless overridden by the <CODE>seqtest</CODE> option,
	the unit of measure is residues.
	If <CODE>seqtest</CODE> is specified,
	the unit of measure is sequences.
		<BR>
		<I>See also:</I>
		<CODE><A HREF="#restest">restest</A></CODE> and
		<CODE><A HREF="#seqtest">seqtest</A></CODE>.
	</TD>
	</TR>
</TABLE>
<HR>

<P>
<I>Last modified: 2005-10-25</I>
</P>

<HR>

<P>
Return to the
<A HREF="http://blast.wustl.edu"><EM>WU&nbsp;BLAST&nbsp;Archives</EM></A> home page
</P>

<P>
<small>
Copyright &copy; 2004-2005 by Warren R. Gish, Saint Louis, Missouri 63108 USA.
All rights reserved.
</small>
</P>

</BODY>
</HTML>

